{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "284e3358",
   "metadata": {},
   "source": [
    "### Install pyenv on WSL\n",
    "\n",
    "```\n",
    "# Clone pyenv\n",
    "curl https://pyenv.run | bash\n",
    "# Add to your shell startup (e.g. ~/.bashrc or ~/.zshrc):\n",
    "export PATH=\"$HOME/.pyenv/bin:$PATH\"\n",
    "eval \"$(pyenv init --path)\"\n",
    "eval \"$(pyenv init -)\"\n",
    "eval \"$(pyenv virtualenv-init -)\"\n",
    "# Then reload:\n",
    "exec $SHELL\n",
    "```\n",
    "\n",
    "### Build Python 3.10.12\n",
    "\n",
    "```\n",
    "pyenv install 3.10.12\n",
    "# Verify:\n",
    "pyenv versions\n",
    "```\n",
    "\n",
    "### Create & activate a pyenv-venv\n",
    "\n",
    "```\n",
    "pyenv virtualenv 3.10.12 yakuza_venv\n",
    "pyenv activate yakuza_venv\n",
    "```\n",
    "\n",
    "### Install Jupyter & ipykernel to work with Jypyter Notebooks\n",
    "\n",
    "```\n",
    "pip install jupyter ipykernel\n",
    "```\n",
    "\n",
    "### Install necessary libraries\n",
    "\n",
    "```\n",
    "pip install numpy pandas torch Bio transformers fair-esm xgboost scikit-learn\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f3b642c",
   "metadata": {},
   "source": [
    "### Download Serratia bacterial genomes from NCBI\n",
    "\n",
    "```\n",
    "pip install ncbi-genome-download\n",
    "ncbi-genome-download \\\n",
    "    --section genbank \\\n",
    "    --formats fasta \\\n",
    "    --assembly-levels complete \\\n",
    "    --species-taxid 615,82996 \\\n",
    "    --output-folder ./data/ncbi_bacteria \\\n",
    "    bacteria\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c4f77b6",
   "metadata": {},
   "source": [
    "### pharokka installation\n",
    "#### Using pip:\n",
    "```\n",
    "pip install pharokka PHANOTATE\n",
    "git clone https://github.com/gbouras13/pharokka.git\n",
    "```\n",
    "\n",
    "#### Alternatively with conda:\n",
    "```\n",
    "sudo apt install -y wget git build-essential hmmer barrnap trnascan-se\n",
    "wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
    "bash Miniconda3-latest-Linux-x86_64.sh\n",
    "```\n",
    "#### Restart terminal to finalize conda installation\n",
    "```\n",
    "conda env create -f pharokka/environment.yml\n",
    "conda activate pharokka_env\n",
    "```\n",
    "\n",
    "### Install pharokka databases:\n",
    "```\n",
    "./pharokka/bin/install_databases.py\n",
    "```\n",
    "\n",
    "#### MMseqs2:\n",
    "```\n",
    "wget https://github.com/soedinglab/MMseqs2/releases/download/13-45111/mmseqs-linux-avx2.tar.gz\n",
    "tar xvfz mmseqs-linux-avx2.tar.gz\n",
    "```\n",
    "\n",
    "#### MinCED:\n",
    "```\n",
    "sudo apt install openjdk-11-jre\n",
    "sudo apt install openjdk-11-jdk-headless\n",
    "git clone https://github.com/ctSkennerton/minced.git\n",
    "cd minced\n",
    "make\n",
    "sudo cp minced.jar /usr/local/bin/\n",
    "sudo cp minced /usr/local/bin/\n",
    "cd ..\n",
    "```\n",
    "\n",
    "#### ARAGORN:\n",
    "```\n",
    "mkdir aragorn; cd aragorn\n",
    "wget https://anaconda.org/bioconda/aragorn/1.2.41/download/linux-64/aragorn-1.2.41-h031d066_3.tar.bz2\n",
    "tar -xvjf aragorn-1.2.41-h031d066_3.tar.bz2\n",
    "sudo cp bin/aragorn /usr/local/bin/\n",
    "cd ..\n",
    "```\n",
    "\n",
    "#### Mash:\n",
    "```\n",
    "wget https://github.com/marbl/Mash/releases/download/v2.3/mash-Linux64-v2.3.tar\n",
    "tar -xvf mash-Linux64-v2.3.tar\n",
    "sudo cp mash-Linux64-v2.3/mash /usr/local/bin/\n",
    "```\n",
    "\n",
    "#### Dnaapler:\n",
    "```\n",
    "pip3 install dnaapler\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba71181f",
   "metadata": {},
   "source": [
    "### Install RBPdetect_v4_ESMfinetuned\n",
    "\n",
    "```\n",
    "sudo apt install unzip\n",
    "wget 'https://zenodo.org/records/14810759/files/RBPdetect_v4_ESMfine.zip?download=1' -O 'RBPdetect_v4_ESMfine.zip'\n",
    "unzip ./RBPdetect_v4_ESMfine.zip -d .\n",
    "rm -f RBPdetect_v4_ESMfine.zip\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c60459c",
   "metadata": {},
   "source": [
    "### Bakta\n",
    "#### Using conda (recommended):\n",
    "```\n",
    "conda install -c conda-forge -c bioconda bakta\n",
    "```\n",
    "\n",
    "#### Using pip (must install 3rd party dependencies manually):\n",
    "```\n",
    "pip install bakta\n",
    "```\n",
    "\n",
    "### AMRFinderPlus:\n",
    "```\n",
    "mkdir AMRFinderPlus; cd AMRFinderPlus\n",
    "wget https://github.com/ncbi/amr/releases/download/amrfinder_v4.0.3/amrfinder_binaries_v4.0.3.tar.gz\n",
    "tar -xvzf amrfinder_binaries_v4.0.3.tar.gz\n",
    "cd ..\n",
    "sudo cp ./AMRFinderPlus/* /usr/local/bin/\n",
    "```\n",
    "\n",
    "### PILER-CR:\n",
    "```\n",
    "wget http://www.drive5.com/pilercr/pilercr1.06.tar.gz\n",
    "tar -xvzf pilercr1.06.tar.gz\n",
    "cd pilercr1.06\n",
    "make\n",
    "sudo cp ./pilercr /usr/local/bin/\n",
    "cd ..\n",
    "```\n",
    "\n",
    "### Diamond:\n",
    "```\n",
    "wget https://github.com/bbuchfink/diamond/releases/download/v2.1.10/diamond-linux64.tar.gz\n",
    "tar -xvzf diamond-linux64.tar.gz\n",
    "sudo mv diamond /usr/local/bin/\n",
    "```\n",
    "\n",
    "### Blast+ (update from 2.12.0 to 2.14.0):\n",
    "```\n",
    "sudo apt remove -y ncbi-blast+\n",
    "wget https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.14.0/ncbi-blast-2.14.0+-x64-linux.tar.gz\n",
    "tar -xvzf ncbi-blast-2.14.0+-x64-linux.tar.gz\n",
    "sudo mv ncbi-blast-2.14.0+/bin/* /usr/local/bin/\n",
    "```\n",
    "\n",
    "## Install full database (~30GB):\n",
    "```\n",
    "mkdir bakta_db\n",
    "bakta_db download --output ./bakta_db --type full\n",
    "```\n",
    "\n",
    "### Run against a single genome .fasta:\n",
    "```\n",
    "mkdir ./data/bakta_test\n",
    "bakta --db ./bakta_db/db --output ./data/bakta_test --force ./data/bacteria_genomes/33KP-HG.fasta\n",
    "bakta --db ./bakta_db/db \\\n",
    "      --output ./data/bakta_test \\\n",
    "      --force \\\n",
    "      --skip-trna \\\n",
    "      --skip-tmrna \\\n",
    "      --skip-rrna \\\n",
    "      --skip-ncrna \\\n",
    "      --skip-ncrna-region \\\n",
    "      --skip-crispr \\\n",
    "      --skip-pseudo \\\n",
    "      --skip-sorf \\\n",
    "      --skip-gap \\\n",
    "      --skip-plot \\\n",
    "      ./data/bacteria_genomes/33KP-HG.fasta\n",
    "```\n",
    "\n",
    "### In general:\n",
    "#### For training:\n",
    "```\n",
    "bakta --db ./bakta_db/db \\\n",
    "      --output <bakta_output_path> \\\n",
    "      --force \\\n",
    "      --skip-trna \\\n",
    "      --skip-tmrna \\\n",
    "      --skip-rrna \\\n",
    "      --skip-ncrna \\\n",
    "      --skip-ncrna-region \\\n",
    "      --skip-crispr \\\n",
    "      --skip-pseudo \\\n",
    "      --skip-sorf \\\n",
    "      --skip-gap \\\n",
    "      --skip-plot \\\n",
    "      <serratia_genome>.fasta\n",
    "```\n",
    "#### For inference:\n",
    "```\n",
    "bakta --db <db_path> \\\n",
    "      --output <bakta_output_path> \\\n",
    "      --force \\\n",
    "      <serratia_genome>.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db855fb8",
   "metadata": {},
   "source": [
    "### Ray Tune\n",
    "\n",
    "```\n",
    "pip install ray[tune]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "216be091",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1 - INITIAL SETUP\n",
    "# --------------------------------------------------\n",
    "import numpy as np\n",
    "import pickle\n",
    "from tqdm import tqdm\n",
    "from pathlib import Path\n",
    "from ray import tune\n",
    "from ray.tune.schedulers import ASHAScheduler\n",
    "from functools import partial\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.model_selection import train_test_split, GroupKFold\n",
    "from sklearn.metrics import roc_auc_score, average_precision_score, f1_score, accuracy_score\n",
    "\n",
    "import serraphim_processing as sp\n",
    "import serraphim_features as sf\n",
    "\n",
    "# Project path\n",
    "main_path = '/home/snt/SerraPHIM'\n",
    "\n",
    "# General paths\n",
    "code_path = main_path + '/code'\n",
    "data_path = main_path + '/data'\n",
    "results_path = main_path + '/results'\n",
    "\n",
    "# Bacterial genome data paths\n",
    "bact_genomes_path = data_path + '/bacteria_genomes'\n",
    "ncbi_bact_path = data_path + '/ncbi_bacteria/genbank/bacteria'\n",
    "\n",
    "# Phage genome data and metadata paths\n",
    "phage_genomes_path = data_path + '/phage_genomes'\n",
    "phagescope_multiFASTA_path = data_path + '/PhageScope_data/PhageScope_virulent_complete-HQ_sequences.fasta'\n",
    "\n",
    "phagescope_metadata_path = data_path + '/PhageScope_metadata'\n",
    "metadata_file = 'serratia_complete-HQ_virulent_phage_metadata.tsv'\n",
    "\n",
    "data_suffix = '_training'\n",
    "\n",
    "# Software and database paths\n",
    "phl_path = '/home/snt/PhageHostLearn-main'\n",
    "pharokka_executable_path = phl_path + '/pharokka/bin/pharokka.py'\n",
    "pharokka_db_path = phl_path + '/pharokka/databases'\n",
    "\n",
    "RBP_detect_model_path = main_path + '/RBPdetect_v4_ESMfine'\n",
    "\n",
    "bakta_db_path = phl_path + '/bakta_db/db'\n",
    "bakta_results_path = data_path + '/bakta_results'\n",
    "\n",
    "# Paths for testing purposes:\n",
    "phage_genomes_sample_path = phage_genomes_path + '/sample'\n",
    "bact_genomes_sample_path = bact_genomes_path + '/sample'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ed90fe8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bacterial genomes' directory is not empty. Moving on...\n"
     ]
    }
   ],
   "source": [
    "# 2 - DATA PROCESSING\n",
    "# --------------------------------------------------\n",
    "\n",
    "# Check if the bacterial genomes' directory exists and is empty\n",
    "if not Path(bact_genomes_path).exists() or not any(Path(bact_genomes_path).iterdir()):\n",
    "    print(\"Bacterial genomes' directory is empty. Processing NCBI FASTA files...\")\n",
    "    sp.process_ncbi_fasta(\n",
    "        input_root=ncbi_bact_path,\n",
    "        output_root=bact_genomes_path,\n",
    "        combine_plasmids=\"True\"\n",
    "    )\n",
    "    # combine_plasmids=\"True\" -> 284 FASTA files saved (plasmid-only FNA files skipped)\n",
    "    # combine_plasmids=\"False\" -> 578 FASTA files saved\n",
    "else:\n",
    "    print(\"Bacterial genomes' directory is not empty. Moving on...\")\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7e10e66f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metadata file already exists: serratia_complete-HQ_virulent_phage_metadata.tsv\n",
      "Phage genomes' directory is not empty. Moving on...\n"
     ]
    }
   ],
   "source": [
    "# Create or load phage metadata\n",
    "phage_metadata = sp.load_or_create_curated_phage_metadata(\n",
    "    phagescope_metadata_path,\n",
    "    metadata_file\n",
    ")\n",
    "\n",
    "# Split the multi-FASTA file manually created using PhageScope\n",
    "if not Path(phage_genomes_path).exists() or not any(Path(phage_genomes_path).iterdir()):\n",
    "    print(\"Phage genomes' directory is empty. Processing PhageScope multi-FASTA file...\")\n",
    "    sp.split_fasta_by_phage_id(\n",
    "        phagescope_multiFASTA_path,\n",
    "        phage_genomes_path\n",
    "    )\n",
    "else:\n",
    "    print(\"Phage genomes' directory is not empty. Moving on...\")\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6e4963b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Interaction matrix already exists at: /home/snt/SerraPHIM/data/interactions/phage_host_interactions_training.csv\n"
     ]
    }
   ],
   "source": [
    "# Derive interaction matrix\n",
    "interaction_filepath = data_path + f'/interactions/phage_host_interactions{data_suffix}.csv'\n",
    "\n",
    "phage_host_interaction_matrix = sp.derive_phage_host_interaction_matrix(\n",
    "    bact_genomes_path,\n",
    "    phage_metadata,\n",
    "    interaction_filepath\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b96c796",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/5 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "2025-07-25 15:51:20.958 | INFO     | input_commands:instantiate_dirs:190 - Removing output directory /home/snt/SerraPHIM/data/pharokka_results/uvig_32658 as -f or --force was specified.\n",
      "2025-07-25 15:51:20.963 | INFO     | __main__:main:95 - Starting Pharokka v1.7.5\n",
      "2025-07-25 15:51:20.963 | INFO     | __main__:main:96 - Command executed: Namespace(infile='/home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta', outdir='/home/snt/SerraPHIM/data/pharokka_results/uvig_32658', database='/home/snt/PhageHostLearn-main/pharokka/databases', threads='16', force=True, prefix='Default', locustag='Default', gene_predictor='default', meta=False, split=False, coding_table='11', evalue='1E-05', fast=False, mmseqs2_only=False, meta_hmm=False, dnaapler=False, custom_hmm='', genbank=False, terminase=False, terminase_strand='nothing', terminase_start='nothing', skip_extra_annotations=False, skip_mash=False, minced_args='', mash_distance=0.2, trna_scan_model='general', citation=False)\n",
      "2025-07-25 15:51:20.963 | INFO     | __main__:main:97 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2025-07-25 15:51:20.963 | INFO     | __main__:main:98 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2025-07-25 15:51:20.964 | INFO     | __main__:main:100 - Checking database installation in /home/snt/PhageHostLearn-main/pharokka/databases.\n",
      "2025-07-25 15:51:20.965 | INFO     | __main__:main:103 - All databases have been successfully checked.\n",
      "2025-07-25 15:51:20.966 | INFO     | __main__:main:119 - Checking dependencies.\n",
      "2025-07-25 15:51:21.121 | INFO     | input_commands:check_dependencies:390 - Phanotate version found is v1.6.7\n",
      "2025-07-25 15:51:21.121 | INFO     | input_commands:check_dependencies:399 - Phanotate version is ok.\n",
      "2025-07-25 15:51:21.141 | INFO     | input_commands:check_dependencies:420 - MMseqs2 version found is 45111b641859ed0ddd875b94d6fd1aef1a675b7e\n",
      "2025-07-25 15:51:21.141 | INFO     | input_commands:check_dependencies:435 - MMseqs2 version is ok.\n",
      "2025-07-25 15:51:21.237 | INFO     | input_commands:check_dependencies:459 - tRNAscan-SE version found is v2.0.12\n",
      "2025-07-25 15:51:21.237 | INFO     | input_commands:check_dependencies:470 - tRNAscan-SE version is ok.\n",
      "2025-07-25 15:51:21.386 | INFO     | input_commands:check_dependencies:494 - MinCED version found is v0.4.2\n",
      "2025-07-25 15:51:21.387 | INFO     | input_commands:check_dependencies:505 - MinCED version is ok.\n",
      "2025-07-25 15:51:21.389 | INFO     | input_commands:check_dependencies:531 - ARAGORN version found is v1.2.41\n",
      "2025-07-25 15:51:21.390 | INFO     | input_commands:check_dependencies:542 - ARAGORN version is ok.\n",
      "2025-07-25 15:51:21.395 | INFO     | input_commands:check_dependencies:559 - mash version found is v2.3\n",
      "2025-07-25 15:51:21.395 | INFO     | input_commands:check_dependencies:566 - mash version is ok.\n",
      "2025-07-25 15:51:21.820 | INFO     | input_commands:check_dependencies:585 - Dnaapler version found is v1.2.0\n",
      "2025-07-25 15:51:21.820 | INFO     | input_commands:check_dependencies:592 - Dnaapler version is ok.\n",
      "2025-07-25 15:51:21.820 | INFO     | input_commands:check_dependencies:604 - Pyrodigal version is v3.6.3\n",
      "2025-07-25 15:51:21.820 | INFO     | input_commands:check_dependencies:605 - Pyrodigal version is ok.\n",
      "2025-07-25 15:51:21.820 | INFO     | input_commands:check_dependencies:617 - Pyrodigal_gv version is v0.3.2\n",
      "2025-07-25 15:51:21.820 | INFO     | input_commands:check_dependencies:618 - Pyrodigal_gv version is ok.\n",
      "2025-07-25 15:51:21.838 | INFO     | input_commands:validate_fasta:237 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta.\n",
      "2025-07-25 15:51:21.838 | INFO     | input_commands:validate_fasta:239 - Input /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta is in FASTA format.\n",
      "2025-07-25 15:51:21.838 | INFO     | input_commands:validate_fasta:244 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta for duplicate FASTA headers.\n",
      "2025-07-25 15:51:21.838 | INFO     | input_commands:validate_fasta:246 - All headers in /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta are unique.\n",
      "2025-07-25 15:51:21.838 | INFO     | input_commands:validate_gene_predictor:275 - Phanotate will be used for gene prediction.\n",
      "2025-07-25 15:51:21.838 | INFO     | __main__:main:280 - Running Phanotate.\n",
      "2025-07-25 15:51:21.840 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta ...\n",
      "2025-07-25 15:51:23.498 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta\n",
      "2025-07-25 15:51:23.498 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta ...\n",
      "2025-07-25 15:51:25.093 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta\n",
      "2025-07-25 15:51:25.101 | INFO     | __main__:main:309 - Starting tRNA-scanSE.\n",
      "2025-07-25 15:51:25.101 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta ...\n",
      "2025-07-25 15:51:25.282 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta\n",
      "2025-07-25 15:51:25.283 | INFO     | processes:run_minced:827 - Running MinCED.\n",
      "2025-07-25 15:51:25.283 | INFO     | external_tools:run:50 - Started running minced /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/pharokka_minced.gff ...\n",
      "2025-07-25 15:51:25.365 | INFO     | external_tools:run:52 - Done running minced /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/pharokka_minced.gff\n",
      "2025-07-25 15:51:25.365 | INFO     | processes:run_aragorn:859 - Running Aragorn.\n",
      "2025-07-25 15:51:25.366 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta ...\n",
      "2025-07-25 15:51:25.370 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta\n",
      "2025-07-25 15:51:25.371 | INFO     | __main__:main:317 - Starting MMseqs2.\n",
      "2025-07-25 15:51:25.371 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on PHROG Database.\n",
      "2025-07-25 15:51:25.371 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/target_dir/target_seqs ...\n",
      "2025-07-25 15:51:25.384 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/target_dir/target_seqs\n",
      "2025-07-25 15:51:25.385 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/tmp_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:53:01.417 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/tmp_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:53:01.418 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/mmseqs_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:53:01.459 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/mmseqs_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:53:01.459 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on CARD Database.\n",
      "2025-07-25 15:53:01.460 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_target_dir/target_seqs ...\n",
      "2025-07-25 15:53:01.469 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_target_dir/target_seqs\n",
      "2025-07-25 15:53:01.470 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:53:15.481 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:53:15.482 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:53:15.500 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/CARD_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:53:15.501 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on VFDB Database.\n",
      "2025-07-25 15:53:15.502 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB_target_dir/target_seqs ...\n",
      "2025-07-25 15:53:15.512 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB_target_dir/target_seqs\n",
      "2025-07-25 15:53:15.513 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:53:39.798 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:53:39.799 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/vfdb_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:53:39.816 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/vfdb_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:53:39.817 | INFO     | __main__:main:348 - Running PyHMMER on PHROGs.\n",
      "2025-07-25 15:53:47.358 | INFO     | __main__:main:364 - Post Processing Output.\n",
      "2025-07-25 15:53:47.364 | INFO     | post_processing:create_mmseqs_tophits:2146 - Processing MMseqs2 outputs.\n",
      "2025-07-25 15:53:47.364 | INFO     | post_processing:create_mmseqs_tophits:2147 - Processing PHROGs output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2169: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"mmseqs_eVal\"))\n",
      "2025-07-25 15:53:47.440 | INFO     | post_processing:process_vfdb_results:2351 - Processing VFDB output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2375: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"vfdb_eVal\"))\n",
      "2025-07-25 15:53:47.446 | INFO     | post_processing:process_vfdb_results:2437 - 0 VFDB virulence factors identified.\n",
      "2025-07-25 15:53:47.446 | INFO     | post_processing:process_card_results:2455 - Processing CARD output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2475: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"CARD_eVal\"))\n",
      "2025-07-25 15:53:47.451 | INFO     | post_processing:process_card_results:2578 - 0 CARD AMR genes identified.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:1110: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  total_gff = total_gff.groupby(\n",
      "2025-07-25 15:53:47.479 | INFO     | __main__:main:435 - Converting gff to genbank.\n",
      "2025-07-25 15:53:47.486 | INFO     | post_processing:extract_terl:1982 - Terminase large subunit found.\n",
      "2025-07-25 15:53:47.487 | INFO     | __main__:main:453 - Finding the closest match for each contig in INPHARED using mash.\n",
      "2025-07-25 15:53:47.488 | INFO     | external_tools:run:50 - Started running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta -o /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/input_mash_sketch.msh -i ...\n",
      "2025-07-25 15:53:47.492 | INFO     | external_tools:run:52 - Done running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/uvig_32658.fasta -o /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/input_mash_sketch.msh -i\n",
      "2025-07-25 15:53:47.492 | INFO     | external_tools:run_to_stdout:59 - Started running mash dist /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i ...\n",
      "2025-07-25 15:53:48.336 | INFO     | external_tools:run_to_stdout:61 - Done running mash dist /home/snt/SerraPHIM/data/pharokka_results/uvig_32658/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i\n",
      "2025-07-25 15:53:48.429 | INFO     | __main__:main:473 - Pharokka has finished.\n",
      "2025-07-25 15:53:48.429 | INFO     | __main__:main:474 - Elapsed time: 147.47 seconds\n",
      "2025-07-25 15:53:48.429 | INFO     | __main__:main:476 - If you use Pharokka in your research, please cite:\n",
      "2025-07-25 15:53:48.429 | INFO     | __main__:main:477 - George Bouras, Roshan Nepal, Ghais Houtak, Alkis James Psaltis, Peter-John Wormald, Sarah Vreugde.\n",
      "2025-07-25 15:53:48.429 | INFO     | __main__:main:480 - Pharokka: a fast scalable bacteriophage annotation tool.\n",
      "2025-07-25 15:53:48.429 | INFO     | __main__:main:481 - Bioinformatics, Volume 39, Issue 1, January 2023, btac776.\n",
      "2025-07-25 15:53:48.429 | INFO     | __main__:main:482 - https://doi.org/10.1093/bioinformatics/btac776.\n",
      "2025-07-25 15:53:48.429 | INFO     | __main__:main:483 - You should also cite the full list of tools Pharokka uses, which can be found at https://github.com/gbouras13/pharokka#citation.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 20%|██        | 1/5 [02:28<09:54, 148.59s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "2025-07-25 15:53:49.243 | INFO     | input_commands:instantiate_dirs:190 - Removing output directory /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631 as -f or --force was specified.\n",
      "2025-07-25 15:53:49.253 | INFO     | __main__:main:95 - Starting Pharokka v1.7.5\n",
      "2025-07-25 15:53:49.253 | INFO     | __main__:main:96 - Command executed: Namespace(infile='/home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta', outdir='/home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631', database='/home/snt/PhageHostLearn-main/pharokka/databases', threads='16', force=True, prefix='Default', locustag='Default', gene_predictor='default', meta=False, split=False, coding_table='11', evalue='1E-05', fast=False, mmseqs2_only=False, meta_hmm=False, dnaapler=False, custom_hmm='', genbank=False, terminase=False, terminase_strand='nothing', terminase_start='nothing', skip_extra_annotations=False, skip_mash=False, minced_args='', mash_distance=0.2, trna_scan_model='general', citation=False)\n",
      "2025-07-25 15:53:49.253 | INFO     | __main__:main:97 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2025-07-25 15:53:49.253 | INFO     | __main__:main:98 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2025-07-25 15:53:49.253 | INFO     | __main__:main:100 - Checking database installation in /home/snt/PhageHostLearn-main/pharokka/databases.\n",
      "2025-07-25 15:53:49.253 | INFO     | __main__:main:103 - All databases have been successfully checked.\n",
      "2025-07-25 15:53:49.254 | INFO     | __main__:main:119 - Checking dependencies.\n",
      "2025-07-25 15:53:49.349 | INFO     | input_commands:check_dependencies:390 - Phanotate version found is v1.6.7\n",
      "2025-07-25 15:53:49.349 | INFO     | input_commands:check_dependencies:399 - Phanotate version is ok.\n",
      "2025-07-25 15:53:49.351 | INFO     | input_commands:check_dependencies:420 - MMseqs2 version found is 45111b641859ed0ddd875b94d6fd1aef1a675b7e\n",
      "2025-07-25 15:53:49.351 | INFO     | input_commands:check_dependencies:435 - MMseqs2 version is ok.\n",
      "2025-07-25 15:53:49.393 | INFO     | input_commands:check_dependencies:459 - tRNAscan-SE version found is v2.0.12\n",
      "2025-07-25 15:53:49.393 | INFO     | input_commands:check_dependencies:470 - tRNAscan-SE version is ok.\n",
      "2025-07-25 15:53:49.433 | INFO     | input_commands:check_dependencies:494 - MinCED version found is v0.4.2\n",
      "2025-07-25 15:53:49.434 | INFO     | input_commands:check_dependencies:505 - MinCED version is ok.\n",
      "2025-07-25 15:53:49.435 | INFO     | input_commands:check_dependencies:531 - ARAGORN version found is v1.2.41\n",
      "2025-07-25 15:53:49.435 | INFO     | input_commands:check_dependencies:542 - ARAGORN version is ok.\n",
      "2025-07-25 15:53:49.437 | INFO     | input_commands:check_dependencies:559 - mash version found is v2.3\n",
      "2025-07-25 15:53:49.437 | INFO     | input_commands:check_dependencies:566 - mash version is ok.\n",
      "2025-07-25 15:53:49.774 | INFO     | input_commands:check_dependencies:585 - Dnaapler version found is v1.2.0\n",
      "2025-07-25 15:53:49.774 | INFO     | input_commands:check_dependencies:592 - Dnaapler version is ok.\n",
      "2025-07-25 15:53:49.774 | INFO     | input_commands:check_dependencies:604 - Pyrodigal version is v3.6.3\n",
      "2025-07-25 15:53:49.774 | INFO     | input_commands:check_dependencies:605 - Pyrodigal version is ok.\n",
      "2025-07-25 15:53:49.775 | INFO     | input_commands:check_dependencies:617 - Pyrodigal_gv version is v0.3.2\n",
      "2025-07-25 15:53:49.775 | INFO     | input_commands:check_dependencies:618 - Pyrodigal_gv version is ok.\n",
      "2025-07-25 15:53:49.779 | INFO     | input_commands:validate_fasta:237 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta.\n",
      "2025-07-25 15:53:49.779 | INFO     | input_commands:validate_fasta:239 - Input /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta is in FASTA format.\n",
      "2025-07-25 15:53:49.780 | INFO     | input_commands:validate_fasta:244 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta for duplicate FASTA headers.\n",
      "2025-07-25 15:53:49.780 | INFO     | input_commands:validate_fasta:246 - All headers in /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta are unique.\n",
      "2025-07-25 15:53:49.780 | INFO     | input_commands:validate_gene_predictor:275 - Phanotate will be used for gene prediction.\n",
      "2025-07-25 15:53:49.780 | INFO     | __main__:main:280 - Running Phanotate.\n",
      "2025-07-25 15:53:49.782 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta ...\n",
      "2025-07-25 15:53:55.301 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta\n",
      "2025-07-25 15:53:55.302 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta ...\n",
      "2025-07-25 15:54:00.576 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta\n",
      "2025-07-25 15:54:00.586 | INFO     | __main__:main:309 - Starting tRNA-scanSE.\n",
      "2025-07-25 15:54:00.587 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta ...\n",
      "2025-07-25 15:54:00.747 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta\n",
      "2025-07-25 15:54:00.747 | INFO     | processes:run_minced:827 - Running MinCED.\n",
      "2025-07-25 15:54:00.748 | INFO     | external_tools:run:50 - Started running minced /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/pharokka_minced.gff ...\n",
      "2025-07-25 15:54:00.823 | INFO     | external_tools:run:52 - Done running minced /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/pharokka_minced.gff\n",
      "2025-07-25 15:54:00.823 | INFO     | processes:run_aragorn:859 - Running Aragorn.\n",
      "2025-07-25 15:54:00.824 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta ...\n",
      "2025-07-25 15:54:00.831 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta\n",
      "2025-07-25 15:54:00.831 | INFO     | __main__:main:317 - Starting MMseqs2.\n",
      "2025-07-25 15:54:00.831 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on PHROG Database.\n",
      "2025-07-25 15:54:00.833 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/target_dir/target_seqs ...\n",
      "2025-07-25 15:54:00.842 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/target_dir/target_seqs\n",
      "2025-07-25 15:54:00.843 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/tmp_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:55:32.265 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/tmp_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:55:32.266 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/mmseqs_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:55:32.303 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/mmseqs_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:55:32.304 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on CARD Database.\n",
      "2025-07-25 15:55:32.305 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_target_dir/target_seqs ...\n",
      "2025-07-25 15:55:32.315 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_target_dir/target_seqs\n",
      "2025-07-25 15:55:32.316 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:55:47.825 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:55:47.826 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:55:47.846 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/CARD_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:55:47.847 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on VFDB Database.\n",
      "2025-07-25 15:55:47.855 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB_target_dir/target_seqs ...\n",
      "2025-07-25 15:55:47.867 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB_target_dir/target_seqs\n",
      "2025-07-25 15:55:47.869 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:56:15.372 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:56:15.374 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/vfdb_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:56:15.390 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/vfdb_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:56:15.391 | INFO     | __main__:main:348 - Running PyHMMER on PHROGs.\n",
      "2025-07-25 15:56:23.350 | INFO     | __main__:main:364 - Post Processing Output.\n",
      "2025-07-25 15:56:23.357 | INFO     | post_processing:create_mmseqs_tophits:2146 - Processing MMseqs2 outputs.\n",
      "2025-07-25 15:56:23.357 | INFO     | post_processing:create_mmseqs_tophits:2147 - Processing PHROGs output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2169: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"mmseqs_eVal\"))\n",
      "2025-07-25 15:56:23.465 | INFO     | post_processing:process_vfdb_results:2351 - Processing VFDB output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2375: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"vfdb_eVal\"))\n",
      "2025-07-25 15:56:23.471 | INFO     | post_processing:process_vfdb_results:2437 - 0 VFDB virulence factors identified.\n",
      "2025-07-25 15:56:23.471 | INFO     | post_processing:process_card_results:2455 - Processing CARD output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2475: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"CARD_eVal\"))\n",
      "2025-07-25 15:56:23.478 | INFO     | post_processing:process_card_results:2578 - 0 CARD AMR genes identified.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:1110: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  total_gff = total_gff.groupby(\n",
      "2025-07-25 15:56:23.513 | INFO     | __main__:main:435 - Converting gff to genbank.\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/Bio/SeqIO/InsdcIO.py:773: BiopythonWarning: Increasing length of locus line to allow long name. This will result in fields that are not in usual positions.\n",
      "  warnings.warn(\n",
      "2025-07-25 15:56:23.531 | INFO     | post_processing:extract_terl:1982 - Terminase large subunit found.\n",
      "2025-07-25 15:56:23.535 | INFO     | __main__:main:453 - Finding the closest match for each contig in INPHARED using mash.\n",
      "2025-07-25 15:56:23.535 | INFO     | external_tools:run:50 - Started running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/input_mash_sketch.msh -i ...\n",
      "2025-07-25 15:56:23.539 | INFO     | external_tools:run:52 - Done running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631.fasta -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/input_mash_sketch.msh -i\n",
      "2025-07-25 15:56:23.540 | INFO     | external_tools:run_to_stdout:59 - Started running mash dist /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i ...\n",
      "2025-07-25 15:56:24.218 | INFO     | external_tools:run_to_stdout:61 - Done running mash dist /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i\n",
      "2025-07-25 15:56:24.318 | INFO     | __main__:main:473 - Pharokka has finished.\n",
      "2025-07-25 15:56:24.318 | INFO     | __main__:main:474 - Elapsed time: 155.07 seconds\n",
      "2025-07-25 15:56:24.318 | INFO     | __main__:main:476 - If you use Pharokka in your research, please cite:\n",
      "2025-07-25 15:56:24.318 | INFO     | __main__:main:477 - George Bouras, Roshan Nepal, Ghais Houtak, Alkis James Psaltis, Peter-John Wormald, Sarah Vreugde.\n",
      "2025-07-25 15:56:24.318 | INFO     | __main__:main:480 - Pharokka: a fast scalable bacteriophage annotation tool.\n",
      "2025-07-25 15:56:24.318 | INFO     | __main__:main:481 - Bioinformatics, Volume 39, Issue 1, January 2023, btac776.\n",
      "2025-07-25 15:56:24.318 | INFO     | __main__:main:482 - https://doi.org/10.1093/bioinformatics/btac776.\n",
      "2025-07-25 15:56:24.318 | INFO     | __main__:main:483 - You should also cite the full list of tools Pharokka uses, which can be found at https://github.com/gbouras13/pharokka#citation.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 40%|████      | 2/5 [05:04<07:38, 152.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "2025-07-25 15:56:25.396 | INFO     | input_commands:instantiate_dirs:190 - Removing output directory /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089 as -f or --force was specified.\n",
      "2025-07-25 15:56:25.407 | INFO     | __main__:main:95 - Starting Pharokka v1.7.5\n",
      "2025-07-25 15:56:25.407 | INFO     | __main__:main:96 - Command executed: Namespace(infile='/home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta', outdir='/home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089', database='/home/snt/PhageHostLearn-main/pharokka/databases', threads='16', force=True, prefix='Default', locustag='Default', gene_predictor='default', meta=False, split=False, coding_table='11', evalue='1E-05', fast=False, mmseqs2_only=False, meta_hmm=False, dnaapler=False, custom_hmm='', genbank=False, terminase=False, terminase_strand='nothing', terminase_start='nothing', skip_extra_annotations=False, skip_mash=False, minced_args='', mash_distance=0.2, trna_scan_model='general', citation=False)\n",
      "2025-07-25 15:56:25.407 | INFO     | __main__:main:97 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2025-07-25 15:56:25.407 | INFO     | __main__:main:98 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2025-07-25 15:56:25.407 | INFO     | __main__:main:100 - Checking database installation in /home/snt/PhageHostLearn-main/pharokka/databases.\n",
      "2025-07-25 15:56:25.407 | INFO     | __main__:main:103 - All databases have been successfully checked.\n",
      "2025-07-25 15:56:25.408 | INFO     | __main__:main:119 - Checking dependencies.\n",
      "2025-07-25 15:56:25.532 | INFO     | input_commands:check_dependencies:390 - Phanotate version found is v1.6.7\n",
      "2025-07-25 15:56:25.532 | INFO     | input_commands:check_dependencies:399 - Phanotate version is ok.\n",
      "2025-07-25 15:56:25.534 | INFO     | input_commands:check_dependencies:420 - MMseqs2 version found is 45111b641859ed0ddd875b94d6fd1aef1a675b7e\n",
      "2025-07-25 15:56:25.534 | INFO     | input_commands:check_dependencies:435 - MMseqs2 version is ok.\n",
      "2025-07-25 15:56:25.579 | INFO     | input_commands:check_dependencies:459 - tRNAscan-SE version found is v2.0.12\n",
      "2025-07-25 15:56:25.579 | INFO     | input_commands:check_dependencies:470 - tRNAscan-SE version is ok.\n",
      "2025-07-25 15:56:25.633 | INFO     | input_commands:check_dependencies:494 - MinCED version found is v0.4.2\n",
      "2025-07-25 15:56:25.633 | INFO     | input_commands:check_dependencies:505 - MinCED version is ok.\n",
      "2025-07-25 15:56:25.635 | INFO     | input_commands:check_dependencies:531 - ARAGORN version found is v1.2.41\n",
      "2025-07-25 15:56:25.635 | INFO     | input_commands:check_dependencies:542 - ARAGORN version is ok.\n",
      "2025-07-25 15:56:25.637 | INFO     | input_commands:check_dependencies:559 - mash version found is v2.3\n",
      "2025-07-25 15:56:25.637 | INFO     | input_commands:check_dependencies:566 - mash version is ok.\n",
      "2025-07-25 15:56:26.096 | INFO     | input_commands:check_dependencies:585 - Dnaapler version found is v1.2.0\n",
      "2025-07-25 15:56:26.096 | INFO     | input_commands:check_dependencies:592 - Dnaapler version is ok.\n",
      "2025-07-25 15:56:26.096 | INFO     | input_commands:check_dependencies:604 - Pyrodigal version is v3.6.3\n",
      "2025-07-25 15:56:26.096 | INFO     | input_commands:check_dependencies:605 - Pyrodigal version is ok.\n",
      "2025-07-25 15:56:26.096 | INFO     | input_commands:check_dependencies:617 - Pyrodigal_gv version is v0.3.2\n",
      "2025-07-25 15:56:26.096 | INFO     | input_commands:check_dependencies:618 - Pyrodigal_gv version is ok.\n",
      "2025-07-25 15:56:26.101 | INFO     | input_commands:validate_fasta:237 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta.\n",
      "2025-07-25 15:56:26.110 | INFO     | input_commands:validate_fasta:239 - Input /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta is in FASTA format.\n",
      "2025-07-25 15:56:26.111 | INFO     | input_commands:validate_fasta:244 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta for duplicate FASTA headers.\n",
      "2025-07-25 15:56:26.111 | INFO     | input_commands:validate_fasta:246 - All headers in /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta are unique.\n",
      "2025-07-25 15:56:26.111 | INFO     | input_commands:validate_gene_predictor:275 - Phanotate will be used for gene prediction.\n",
      "2025-07-25 15:56:26.111 | INFO     | __main__:main:280 - Running Phanotate.\n",
      "2025-07-25 15:56:26.112 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta ...\n",
      "2025-07-25 15:56:32.590 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta\n",
      "2025-07-25 15:56:32.590 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta ...\n",
      "2025-07-25 15:56:38.874 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta\n",
      "2025-07-25 15:56:38.887 | INFO     | __main__:main:309 - Starting tRNA-scanSE.\n",
      "2025-07-25 15:56:38.888 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta ...\n",
      "2025-07-25 15:56:39.082 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta\n",
      "2025-07-25 15:56:39.083 | INFO     | processes:run_minced:827 - Running MinCED.\n",
      "2025-07-25 15:56:39.083 | INFO     | external_tools:run:50 - Started running minced /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/pharokka_minced.gff ...\n",
      "2025-07-25 15:56:39.177 | INFO     | external_tools:run:52 - Done running minced /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/pharokka_minced.gff\n",
      "2025-07-25 15:56:39.177 | INFO     | processes:run_aragorn:859 - Running Aragorn.\n",
      "2025-07-25 15:56:39.178 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta ...\n",
      "2025-07-25 15:56:39.188 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta\n",
      "2025-07-25 15:56:39.189 | INFO     | __main__:main:317 - Starting MMseqs2.\n",
      "2025-07-25 15:56:39.189 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on PHROG Database.\n",
      "2025-07-25 15:56:39.190 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/target_dir/target_seqs ...\n",
      "2025-07-25 15:56:39.201 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/target_dir/target_seqs\n",
      "2025-07-25 15:56:39.201 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/tmp_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:58:14.800 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/tmp_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:58:14.802 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/mmseqs_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:58:14.839 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/mmseqs_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:58:14.840 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on CARD Database.\n",
      "2025-07-25 15:58:14.841 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_target_dir/target_seqs ...\n",
      "2025-07-25 15:58:14.852 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_target_dir/target_seqs\n",
      "2025-07-25 15:58:14.853 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:58:30.376 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:58:30.378 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:58:30.390 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/CARD_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:58:30.391 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on VFDB Database.\n",
      "2025-07-25 15:58:30.399 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB_target_dir/target_seqs ...\n",
      "2025-07-25 15:58:30.412 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB_target_dir/target_seqs\n",
      "2025-07-25 15:58:30.413 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 15:58:57.359 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 15:58:57.360 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/vfdb_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 15:58:57.380 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/vfdb_results.tsv --full-header --threads 16\n",
      "2025-07-25 15:58:57.381 | INFO     | __main__:main:348 - Running PyHMMER on PHROGs.\n",
      "2025-07-25 15:59:07.049 | INFO     | __main__:main:364 - Post Processing Output.\n",
      "2025-07-25 15:59:07.058 | INFO     | post_processing:create_mmseqs_tophits:2146 - Processing MMseqs2 outputs.\n",
      "2025-07-25 15:59:07.058 | INFO     | post_processing:create_mmseqs_tophits:2147 - Processing PHROGs output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2169: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"mmseqs_eVal\"))\n",
      "2025-07-25 15:59:07.190 | INFO     | post_processing:process_vfdb_results:2351 - Processing VFDB output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2375: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"vfdb_eVal\"))\n",
      "2025-07-25 15:59:07.196 | INFO     | post_processing:process_vfdb_results:2437 - 0 VFDB virulence factors identified.\n",
      "2025-07-25 15:59:07.196 | INFO     | post_processing:process_card_results:2455 - Processing CARD output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2475: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"CARD_eVal\"))\n",
      "2025-07-25 15:59:07.202 | INFO     | post_processing:process_card_results:2578 - 0 CARD AMR genes identified.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:1110: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  total_gff = total_gff.groupby(\n",
      "2025-07-25 15:59:07.239 | INFO     | __main__:main:435 - Converting gff to genbank.\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/Bio/SeqIO/InsdcIO.py:773: BiopythonWarning: Increasing length of locus line to allow long name. This will result in fields that are not in usual positions.\n",
      "  warnings.warn(\n",
      "2025-07-25 15:59:07.259 | INFO     | post_processing:extract_terl:1982 - Terminase large subunit found.\n",
      "2025-07-25 15:59:07.262 | INFO     | __main__:main:453 - Finding the closest match for each contig in INPHARED using mash.\n",
      "2025-07-25 15:59:07.262 | INFO     | external_tools:run:50 - Started running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/input_mash_sketch.msh -i ...\n",
      "2025-07-25 15:59:07.267 | INFO     | external_tools:run:52 - Done running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/IMGVR_UViG_640753048_000002_640753048_640753089.fasta -o /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/input_mash_sketch.msh -i\n",
      "2025-07-25 15:59:07.268 | INFO     | external_tools:run_to_stdout:59 - Started running mash dist /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i ...\n",
      "2025-07-25 15:59:08.058 | INFO     | external_tools:run_to_stdout:61 - Done running mash dist /home/snt/SerraPHIM/data/pharokka_results/IMGVR_UViG_640753048_000002_640753048_640753089/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i\n",
      "2025-07-25 15:59:08.161 | INFO     | __main__:main:473 - Pharokka has finished.\n",
      "2025-07-25 15:59:08.161 | INFO     | __main__:main:474 - Elapsed time: 162.76 seconds\n",
      "2025-07-25 15:59:08.161 | INFO     | __main__:main:476 - If you use Pharokka in your research, please cite:\n",
      "2025-07-25 15:59:08.161 | INFO     | __main__:main:477 - George Bouras, Roshan Nepal, Ghais Houtak, Alkis James Psaltis, Peter-John Wormald, Sarah Vreugde.\n",
      "2025-07-25 15:59:08.161 | INFO     | __main__:main:480 - Pharokka: a fast scalable bacteriophage annotation tool.\n",
      "2025-07-25 15:59:08.161 | INFO     | __main__:main:481 - Bioinformatics, Volume 39, Issue 1, January 2023, btac776.\n",
      "2025-07-25 15:59:08.161 | INFO     | __main__:main:482 - https://doi.org/10.1093/bioinformatics/btac776.\n",
      "2025-07-25 15:59:08.161 | INFO     | __main__:main:483 - You should also cite the full list of tools Pharokka uses, which can be found at https://github.com/gbouras13/pharokka#citation.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 60%|██████    | 3/5 [07:48<05:15, 157.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "2025-07-25 15:59:09.284 | INFO     | input_commands:instantiate_dirs:190 - Removing output directory /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1 as -f or --force was specified.\n",
      "2025-07-25 15:59:09.294 | INFO     | __main__:main:95 - Starting Pharokka v1.7.5\n",
      "2025-07-25 15:59:09.294 | INFO     | __main__:main:96 - Command executed: Namespace(infile='/home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta', outdir='/home/snt/SerraPHIM/data/pharokka_results/NC_021563.1', database='/home/snt/PhageHostLearn-main/pharokka/databases', threads='16', force=True, prefix='Default', locustag='Default', gene_predictor='default', meta=False, split=False, coding_table='11', evalue='1E-05', fast=False, mmseqs2_only=False, meta_hmm=False, dnaapler=False, custom_hmm='', genbank=False, terminase=False, terminase_strand='nothing', terminase_start='nothing', skip_extra_annotations=False, skip_mash=False, minced_args='', mash_distance=0.2, trna_scan_model='general', citation=False)\n",
      "2025-07-25 15:59:09.295 | INFO     | __main__:main:97 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2025-07-25 15:59:09.295 | INFO     | __main__:main:98 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2025-07-25 15:59:09.295 | INFO     | __main__:main:100 - Checking database installation in /home/snt/PhageHostLearn-main/pharokka/databases.\n",
      "2025-07-25 15:59:09.295 | INFO     | __main__:main:103 - All databases have been successfully checked.\n",
      "2025-07-25 15:59:09.295 | INFO     | __main__:main:119 - Checking dependencies.\n",
      "2025-07-25 15:59:09.425 | INFO     | input_commands:check_dependencies:390 - Phanotate version found is v1.6.7\n",
      "2025-07-25 15:59:09.425 | INFO     | input_commands:check_dependencies:399 - Phanotate version is ok.\n",
      "2025-07-25 15:59:09.427 | INFO     | input_commands:check_dependencies:420 - MMseqs2 version found is 45111b641859ed0ddd875b94d6fd1aef1a675b7e\n",
      "2025-07-25 15:59:09.427 | INFO     | input_commands:check_dependencies:435 - MMseqs2 version is ok.\n",
      "2025-07-25 15:59:09.473 | INFO     | input_commands:check_dependencies:459 - tRNAscan-SE version found is v2.0.12\n",
      "2025-07-25 15:59:09.474 | INFO     | input_commands:check_dependencies:470 - tRNAscan-SE version is ok.\n",
      "2025-07-25 15:59:09.523 | INFO     | input_commands:check_dependencies:494 - MinCED version found is v0.4.2\n",
      "2025-07-25 15:59:09.523 | INFO     | input_commands:check_dependencies:505 - MinCED version is ok.\n",
      "2025-07-25 15:59:09.525 | INFO     | input_commands:check_dependencies:531 - ARAGORN version found is v1.2.41\n",
      "2025-07-25 15:59:09.525 | INFO     | input_commands:check_dependencies:542 - ARAGORN version is ok.\n",
      "2025-07-25 15:59:09.526 | INFO     | input_commands:check_dependencies:559 - mash version found is v2.3\n",
      "2025-07-25 15:59:09.526 | INFO     | input_commands:check_dependencies:566 - mash version is ok.\n",
      "2025-07-25 15:59:09.976 | INFO     | input_commands:check_dependencies:585 - Dnaapler version found is v1.2.0\n",
      "2025-07-25 15:59:09.976 | INFO     | input_commands:check_dependencies:592 - Dnaapler version is ok.\n",
      "2025-07-25 15:59:09.976 | INFO     | input_commands:check_dependencies:604 - Pyrodigal version is v3.6.3\n",
      "2025-07-25 15:59:09.976 | INFO     | input_commands:check_dependencies:605 - Pyrodigal version is ok.\n",
      "2025-07-25 15:59:09.976 | INFO     | input_commands:check_dependencies:617 - Pyrodigal_gv version is v0.3.2\n",
      "2025-07-25 15:59:09.976 | INFO     | input_commands:check_dependencies:618 - Pyrodigal_gv version is ok.\n",
      "2025-07-25 15:59:09.987 | INFO     | input_commands:validate_fasta:237 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta.\n",
      "2025-07-25 15:59:09.989 | INFO     | input_commands:validate_fasta:239 - Input /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta is in FASTA format.\n",
      "2025-07-25 15:59:09.989 | INFO     | input_commands:validate_fasta:244 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta for duplicate FASTA headers.\n",
      "2025-07-25 15:59:09.989 | INFO     | input_commands:validate_fasta:246 - All headers in /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta are unique.\n",
      "2025-07-25 15:59:09.989 | INFO     | input_commands:validate_gene_predictor:275 - Phanotate will be used for gene prediction.\n",
      "2025-07-25 15:59:09.989 | INFO     | __main__:main:280 - Running Phanotate.\n",
      "2025-07-25 15:59:09.991 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta ...\n",
      "2025-07-25 15:59:17.075 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta\n",
      "2025-07-25 15:59:17.076 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta ...\n",
      "2025-07-25 15:59:24.186 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta\n",
      "2025-07-25 15:59:24.198 | INFO     | __main__:main:309 - Starting tRNA-scanSE.\n",
      "2025-07-25 15:59:24.198 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta ...\n",
      "2025-07-25 15:59:24.410 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta\n",
      "2025-07-25 15:59:24.411 | INFO     | processes:run_minced:827 - Running MinCED.\n",
      "2025-07-25 15:59:24.411 | INFO     | external_tools:run:50 - Started running minced /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/pharokka_minced.gff ...\n",
      "2025-07-25 15:59:24.502 | INFO     | external_tools:run:52 - Done running minced /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/pharokka_minced.gff\n",
      "2025-07-25 15:59:24.503 | INFO     | processes:run_aragorn:859 - Running Aragorn.\n",
      "2025-07-25 15:59:24.503 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta ...\n",
      "2025-07-25 15:59:24.513 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta\n",
      "2025-07-25 15:59:24.513 | INFO     | __main__:main:317 - Starting MMseqs2.\n",
      "2025-07-25 15:59:24.513 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on PHROG Database.\n",
      "2025-07-25 15:59:24.515 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/target_dir/target_seqs ...\n",
      "2025-07-25 15:59:24.525 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/target_dir/target_seqs\n",
      "2025-07-25 15:59:24.526 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/tmp_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 16:01:04.114 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/tmp_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 16:01:04.116 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/mmseqs_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 16:01:04.151 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/mmseqs_results.tsv --full-header --threads 16\n",
      "2025-07-25 16:01:04.152 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on CARD Database.\n",
      "2025-07-25 16:01:04.158 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_target_dir/target_seqs ...\n",
      "2025-07-25 16:01:04.169 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_target_dir/target_seqs\n",
      "2025-07-25 16:01:04.171 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 16:01:19.949 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 16:01:19.950 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 16:01:19.963 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/CARD_results.tsv --full-header --threads 16\n",
      "2025-07-25 16:01:19.964 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on VFDB Database.\n",
      "2025-07-25 16:01:19.970 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB_target_dir/target_seqs ...\n",
      "2025-07-25 16:01:19.984 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB_target_dir/target_seqs\n",
      "2025-07-25 16:01:19.985 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 16:01:47.018 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 16:01:47.019 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/vfdb_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 16:01:47.033 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/vfdb_results.tsv --full-header --threads 16\n",
      "2025-07-25 16:01:47.034 | INFO     | __main__:main:348 - Running PyHMMER on PHROGs.\n",
      "2025-07-25 16:01:55.359 | INFO     | __main__:main:364 - Post Processing Output.\n",
      "2025-07-25 16:01:55.366 | INFO     | post_processing:create_mmseqs_tophits:2146 - Processing MMseqs2 outputs.\n",
      "2025-07-25 16:01:55.366 | INFO     | post_processing:create_mmseqs_tophits:2147 - Processing PHROGs output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2169: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"mmseqs_eVal\"))\n",
      "2025-07-25 16:01:55.484 | INFO     | post_processing:process_vfdb_results:2351 - Processing VFDB output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2375: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"vfdb_eVal\"))\n",
      "2025-07-25 16:01:55.491 | INFO     | post_processing:process_vfdb_results:2437 - 0 VFDB virulence factors identified.\n",
      "2025-07-25 16:01:55.492 | INFO     | post_processing:process_card_results:2455 - Processing CARD output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2475: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"CARD_eVal\"))\n",
      "2025-07-25 16:01:55.497 | INFO     | post_processing:process_card_results:2578 - 0 CARD AMR genes identified.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:1110: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  total_gff = total_gff.groupby(\n",
      "2025-07-25 16:01:55.533 | INFO     | __main__:main:435 - Converting gff to genbank.\n",
      "2025-07-25 16:01:55.555 | INFO     | __main__:main:453 - Finding the closest match for each contig in INPHARED using mash.\n",
      "2025-07-25 16:01:55.555 | INFO     | external_tools:run:50 - Started running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta -o /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/input_mash_sketch.msh -i ...\n",
      "2025-07-25 16:01:55.559 | INFO     | external_tools:run:52 - Done running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/NC_021563.1.fasta -o /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/input_mash_sketch.msh -i\n",
      "2025-07-25 16:01:55.559 | INFO     | external_tools:run_to_stdout:59 - Started running mash dist /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i ...\n",
      "2025-07-25 16:01:56.234 | INFO     | external_tools:run_to_stdout:61 - Done running mash dist /home/snt/SerraPHIM/data/pharokka_results/NC_021563.1/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i\n",
      "2025-07-25 16:01:56.353 | INFO     | __main__:main:473 - Pharokka has finished.\n",
      "2025-07-25 16:01:56.353 | INFO     | __main__:main:474 - Elapsed time: 167.06 seconds\n",
      "2025-07-25 16:01:56.353 | INFO     | __main__:main:476 - If you use Pharokka in your research, please cite:\n",
      "2025-07-25 16:01:56.353 | INFO     | __main__:main:477 - George Bouras, Roshan Nepal, Ghais Houtak, Alkis James Psaltis, Peter-John Wormald, Sarah Vreugde.\n",
      "2025-07-25 16:01:56.354 | INFO     | __main__:main:480 - Pharokka: a fast scalable bacteriophage annotation tool.\n",
      "2025-07-25 16:01:56.354 | INFO     | __main__:main:481 - Bioinformatics, Volume 39, Issue 1, January 2023, btac776.\n",
      "2025-07-25 16:01:56.354 | INFO     | __main__:main:482 - https://doi.org/10.1093/bioinformatics/btac776.\n",
      "2025-07-25 16:01:56.354 | INFO     | __main__:main:483 - You should also cite the full list of tools Pharokka uses, which can be found at https://github.com/gbouras13/pharokka#citation.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 80%|████████  | 4/5 [10:36<02:41, 161.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:558: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/numpy/_core/getlimits.py:90: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "2025-07-25 16:01:57.417 | INFO     | input_commands:instantiate_dirs:190 - Removing output directory /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1 as -f or --force was specified.\n",
      "2025-07-25 16:01:57.424 | INFO     | __main__:main:95 - Starting Pharokka v1.7.5\n",
      "2025-07-25 16:01:57.424 | INFO     | __main__:main:96 - Command executed: Namespace(infile='/home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta', outdir='/home/snt/SerraPHIM/data/pharokka_results/NC_020083.1', database='/home/snt/PhageHostLearn-main/pharokka/databases', threads='16', force=True, prefix='Default', locustag='Default', gene_predictor='default', meta=False, split=False, coding_table='11', evalue='1E-05', fast=False, mmseqs2_only=False, meta_hmm=False, dnaapler=False, custom_hmm='', genbank=False, terminase=False, terminase_strand='nothing', terminase_start='nothing', skip_extra_annotations=False, skip_mash=False, minced_args='', mash_distance=0.2, trna_scan_model='general', citation=False)\n",
      "2025-07-25 16:01:57.425 | INFO     | __main__:main:97 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2025-07-25 16:01:57.425 | INFO     | __main__:main:98 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2025-07-25 16:01:57.425 | INFO     | __main__:main:100 - Checking database installation in /home/snt/PhageHostLearn-main/pharokka/databases.\n",
      "2025-07-25 16:01:57.425 | INFO     | __main__:main:103 - All databases have been successfully checked.\n",
      "2025-07-25 16:01:57.425 | INFO     | __main__:main:119 - Checking dependencies.\n",
      "2025-07-25 16:01:57.548 | INFO     | input_commands:check_dependencies:390 - Phanotate version found is v1.6.7\n",
      "2025-07-25 16:01:57.548 | INFO     | input_commands:check_dependencies:399 - Phanotate version is ok.\n",
      "2025-07-25 16:01:57.550 | INFO     | input_commands:check_dependencies:420 - MMseqs2 version found is 45111b641859ed0ddd875b94d6fd1aef1a675b7e\n",
      "2025-07-25 16:01:57.550 | INFO     | input_commands:check_dependencies:435 - MMseqs2 version is ok.\n",
      "2025-07-25 16:01:57.595 | INFO     | input_commands:check_dependencies:459 - tRNAscan-SE version found is v2.0.12\n",
      "2025-07-25 16:01:57.595 | INFO     | input_commands:check_dependencies:470 - tRNAscan-SE version is ok.\n",
      "2025-07-25 16:01:57.645 | INFO     | input_commands:check_dependencies:494 - MinCED version found is v0.4.2\n",
      "2025-07-25 16:01:57.645 | INFO     | input_commands:check_dependencies:505 - MinCED version is ok.\n",
      "2025-07-25 16:01:57.647 | INFO     | input_commands:check_dependencies:531 - ARAGORN version found is v1.2.41\n",
      "2025-07-25 16:01:57.647 | INFO     | input_commands:check_dependencies:542 - ARAGORN version is ok.\n",
      "2025-07-25 16:01:57.649 | INFO     | input_commands:check_dependencies:559 - mash version found is v2.3\n",
      "2025-07-25 16:01:57.649 | INFO     | input_commands:check_dependencies:566 - mash version is ok.\n",
      "2025-07-25 16:01:58.104 | INFO     | input_commands:check_dependencies:585 - Dnaapler version found is v1.2.0\n",
      "2025-07-25 16:01:58.104 | INFO     | input_commands:check_dependencies:592 - Dnaapler version is ok.\n",
      "2025-07-25 16:01:58.105 | INFO     | input_commands:check_dependencies:604 - Pyrodigal version is v3.6.3\n",
      "2025-07-25 16:01:58.105 | INFO     | input_commands:check_dependencies:605 - Pyrodigal version is ok.\n",
      "2025-07-25 16:01:58.105 | INFO     | input_commands:check_dependencies:617 - Pyrodigal_gv version is v0.3.2\n",
      "2025-07-25 16:01:58.105 | INFO     | input_commands:check_dependencies:618 - Pyrodigal_gv version is ok.\n",
      "2025-07-25 16:01:58.110 | INFO     | input_commands:validate_fasta:237 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta.\n",
      "2025-07-25 16:01:58.111 | INFO     | input_commands:validate_fasta:239 - Input /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta is in FASTA format.\n",
      "2025-07-25 16:01:58.111 | INFO     | input_commands:validate_fasta:244 - Checking input /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta for duplicate FASTA headers.\n",
      "2025-07-25 16:01:58.111 | INFO     | input_commands:validate_fasta:246 - All headers in /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta are unique.\n",
      "2025-07-25 16:01:58.111 | INFO     | input_commands:validate_gene_predictor:275 - Phanotate will be used for gene prediction.\n",
      "2025-07-25 16:01:58.112 | INFO     | __main__:main:280 - Running Phanotate.\n",
      "2025-07-25 16:01:58.113 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta ...\n",
      "2025-07-25 16:02:33.277 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_out_tmp.fasta /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta\n",
      "2025-07-25 16:02:33.278 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta ...\n",
      "2025-07-25 16:03:09.089 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_out.txt /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta\n",
      "2025-07-25 16:03:09.117 | INFO     | __main__:main:309 - Starting tRNA-scanSE.\n",
      "2025-07-25 16:03:09.117 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta ...\n",
      "2025-07-25 16:03:09.797 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 16 -G -Q -j /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/trnascan_out.gff /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta\n",
      "2025-07-25 16:03:09.797 | INFO     | processes:run_minced:827 - Running MinCED.\n",
      "2025-07-25 16:03:09.798 | INFO     | external_tools:run:50 - Started running minced /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/pharokka_minced.gff ...\n",
      "2025-07-25 16:03:09.899 | INFO     | external_tools:run:52 - Done running minced /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/pharokka_minced_spacers.txt /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/pharokka_minced.gff\n",
      "2025-07-25 16:03:09.900 | INFO     | processes:run_aragorn:859 - Running Aragorn.\n",
      "2025-07-25 16:03:09.900 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta ...\n",
      "2025-07-25 16:03:09.931 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/pharokka_aragorn.txt /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta\n",
      "2025-07-25 16:03:09.932 | INFO     | __main__:main:317 - Starting MMseqs2.\n",
      "2025-07-25 16:03:09.932 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on PHROG Database.\n",
      "2025-07-25 16:03:09.933 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/target_dir/target_seqs ...\n",
      "2025-07-25 16:03:09.945 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/target_dir/target_seqs\n",
      "2025-07-25 16:03:09.946 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/tmp_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 16:04:49.410 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/tmp_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 16:04:49.411 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/mmseqs_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 16:04:49.448 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/phrogs_profile_db /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/mmseqs/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/mmseqs_results.tsv --full-header --threads 16\n",
      "2025-07-25 16:04:49.449 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on CARD Database.\n",
      "2025-07-25 16:04:49.452 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_target_dir/target_seqs ...\n",
      "2025-07-25 16:04:49.463 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_target_dir/target_seqs\n",
      "2025-07-25 16:04:49.464 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 16:05:05.194 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 16:05:05.195 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 16:05:05.214 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/CARD /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/CARD_results.tsv --full-header --threads 16\n",
      "2025-07-25 16:05:05.215 | INFO     | processes:run_mmseqs:685 - Running MMseqs2 on VFDB Database.\n",
      "2025-07-25 16:05:05.222 | INFO     | external_tools:run:50 - Started running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB_target_dir/target_seqs ...\n",
      "2025-07-25 16:05:05.235 | INFO     | external_tools:run:52 - Done running mmseqs createdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/phanotate_aas_tmp.fasta /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB_target_dir/target_seqs\n",
      "2025-07-25 16:05:05.236 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB_dir/ -s 8.5 --threads 16 ...\n",
      "2025-07-25 16:05:32.254 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB_dir/ -s 8.5 --threads 16\n",
      "2025-07-25 16:05:32.255 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/vfdb_results.tsv --full-header --threads 16 ...\n",
      "2025-07-25 16:05:32.269 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /home/snt/PhageHostLearn-main/pharokka/databases/vfdb /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB_target_dir/target_seqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/VFDB/results_mmseqs /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/vfdb_results.tsv --full-header --threads 16\n",
      "2025-07-25 16:05:32.270 | INFO     | __main__:main:348 - Running PyHMMER on PHROGs.\n",
      "2025-07-25 16:05:42.494 | INFO     | __main__:main:364 - Post Processing Output.\n",
      "2025-07-25 16:05:42.508 | INFO     | post_processing:create_mmseqs_tophits:2146 - Processing MMseqs2 outputs.\n",
      "2025-07-25 16:05:42.508 | INFO     | post_processing:create_mmseqs_tophits:2147 - Processing PHROGs output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2169: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"mmseqs_eVal\"))\n",
      "2025-07-25 16:05:42.710 | INFO     | post_processing:process_vfdb_results:2351 - Processing VFDB output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2375: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"vfdb_eVal\"))\n",
      "2025-07-25 16:05:42.715 | INFO     | post_processing:process_vfdb_results:2437 - 0 VFDB virulence factors identified.\n",
      "2025-07-25 16:05:42.716 | INFO     | post_processing:process_card_results:2455 - Processing CARD output.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:2475: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda group: group.nsmallest(1, \"CARD_eVal\"))\n",
      "2025-07-25 16:05:42.721 | INFO     | post_processing:process_card_results:2578 - 0 CARD AMR genes identified.\n",
      "/home/snt/PhageHostLearn-main/pharokka/bin/post_processing.py:1110: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  total_gff = total_gff.groupby(\n",
      "2025-07-25 16:05:42.785 | INFO     | __main__:main:435 - Converting gff to genbank.\n",
      "2025-07-25 16:05:42.829 | INFO     | post_processing:extract_terl:1982 - Terminase large subunit found.\n",
      "2025-07-25 16:05:42.836 | INFO     | __main__:main:453 - Finding the closest match for each contig in INPHARED using mash.\n",
      "2025-07-25 16:05:42.837 | INFO     | external_tools:run:50 - Started running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta -o /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/input_mash_sketch.msh -i ...\n",
      "2025-07-25 16:05:42.843 | INFO     | external_tools:run:52 - Done running mash sketch /home/snt/SerraPHIM/data/phage_genomes/sample/NC_020083.1.fasta -o /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/input_mash_sketch.msh -i\n",
      "2025-07-25 16:05:42.843 | INFO     | external_tools:run_to_stdout:59 - Started running mash dist /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i ...\n",
      "2025-07-25 16:05:43.499 | INFO     | external_tools:run_to_stdout:61 - Done running mash dist /home/snt/SerraPHIM/data/pharokka_results/NC_020083.1/input_mash_sketch.msh /home/snt/PhageHostLearn-main/pharokka/databases/1Aug2023_genomes.fa.msh -d 0.2 -i\n",
      "2025-07-25 16:05:43.612 | INFO     | __main__:main:473 - Pharokka has finished.\n",
      "2025-07-25 16:05:43.612 | INFO     | __main__:main:474 - Elapsed time: 226.19 seconds\n",
      "2025-07-25 16:05:43.612 | INFO     | __main__:main:476 - If you use Pharokka in your research, please cite:\n",
      "2025-07-25 16:05:43.612 | INFO     | __main__:main:477 - George Bouras, Roshan Nepal, Ghais Houtak, Alkis James Psaltis, Peter-John Wormald, Sarah Vreugde.\n",
      "2025-07-25 16:05:43.612 | INFO     | __main__:main:480 - Pharokka: a fast scalable bacteriophage annotation tool.\n",
      "2025-07-25 16:05:43.612 | INFO     | __main__:main:481 - Bioinformatics, Volume 39, Issue 1, January 2023, btac776.\n",
      "2025-07-25 16:05:43.612 | INFO     | __main__:main:482 - https://doi.org/10.1093/bioinformatics/btac776.\n",
      "2025-07-25 16:05:43.612 | INFO     | __main__:main:483 - You should also cite the full list of tools Pharokka uses, which can be found at https://github.com/gbouras13/pharokka#citation.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5/5 [14:23<00:00, 172.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pharokka processing complete.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Run pharokka\n",
    "sp.pharokka_processing(\n",
    "    data_path, \n",
    "    phage_genomes_path, # phage_genomes_path or phage_genomes_sample_path\n",
    "    pharokka_executable_path, \n",
    "    pharokka_db_path, \n",
    "    data_suffix,\n",
    "    threads=16  # 128 cores in yakuza server\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbf58bec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using device: cuda\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing phages:   0%|          | 0/5 [00:00<?, ?it/s]Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\n",
      "Inference for uvig_32658: 100%|██████████| 18/18 [00:00<00:00, 26.70it/s]\n",
      "Inference for IMGVR_UViG_640753048_000002_640753048_640753089: 100%|██████████| 93/93 [00:01<00:00, 69.34it/s]\n",
      "Inference for NC_020083.1: 100%|██████████| 228/228 [00:03<00:00, 63.66it/s]\n",
      "Inference for IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631: 100%|██████████| 74/74 [00:01<00:00, 69.43it/s]\n",
      "Inference for NC_021563.1: 100%|██████████| 77/77 [00:01<00:00, 68.51it/s]\n",
      "Processing phages: 100%|██████████| 5/5 [00:07<00:00,  1.56s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RBP detection complete. Results saved to 'RBPbase_training.csv'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Run RBP detection using PhageRBPdetect_v4\n",
    "# WARNING: No GPUs in yakuza server!\n",
    "sp.process_and_detect_rbps_v4(\n",
    "    data_path, \n",
    "    RBP_detect_model_path, \n",
    "    data_suffix,\n",
    "    gpu_index=0,\n",
    "    threads=16  # 128 cores in yakuza server\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "435c15ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing genomes:   0%|          | 0/5 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command: bakta --db /home/snt/PhageHostLearn-main/bakta_db/db --output /home/snt/SerraPHIM/data/bakta_results/CP100764.1_Serratia_plymuthica_strain_SASK2000_chromosome --force --threads 16 --skip-trna --skip-tmrna --skip-rrna --skip-ncrna --skip-ncrna-region --skip-crispr --skip-pseudo --skip-sorf --skip-gap --skip-plot /home/snt/SerraPHIM/data/bacteria_genomes/sample/CP100764.1_Serratia_plymuthica_strain_SASK2000_chromosome.fasta\n",
      "Parse genome sequences...\n",
      "\timported: 1\n",
      "\tfiltered & revised: 1\n",
      "\tchromosomes: 1\n",
      "\n",
      "Start annotation...\n",
      "skip tRNA prediction...\n",
      "skip tmRNA prediction...\n",
      "skip rRNA prediction...\n",
      "skip ncRNA prediction...\n",
      "skip ncRNA region prediction...\n",
      "skip CRISPR array prediction...\n",
      "predict & annotate CDSs...\n",
      "\tpredicted: 5041 \n",
      "\tdiscarded spurious: 1\n",
      "\trevised translational exceptions: 0\n",
      "\tdetected IPSs: 4764\n",
      "\tfound PSCs: 138\n",
      "\tfound PSCCs: 106\n",
      "\tlookup annotations...\n",
      "\tconduct expert systems...\n",
      "\t\tamrfinder: 7\n",
      "\t\tprotein sequences: 50\n",
      "\tcombine annotations and mark hypotheticals...\n",
      "\tanalyze hypothetical proteins: 218\n",
      "\t\tdetected Pfam hits: 16 \n",
      "\t\tcalculated proteins statistics\n",
      "\trevise special cases...\n",
      "skip sORF prediction...\n",
      "skip gap annotation...\n",
      "detect oriCs/oriVs...\n",
      "\tfound: 1\n",
      "detect oriTs...\n",
      "\tfound: 0\n",
      "apply feature overlap filters...\n",
      "select features and create locus tags...\n",
      "\tselected: 5041\n",
      "improve annotations...\n",
      "\trevised gene symbols: 36\n",
      "\n",
      "Genome statistics:\n",
      "\tGenome size: 5,442,880 bp\n",
      "\tContigs/replicons: 1\n",
      "\tGC: 56.0 %\n",
      "\tN50: 5,442,880\n",
      "\tN90: 5,442,880\n",
      "\tN ratio: 0.0 %\n",
      "\tcoding density: 87.0 %\n",
      "\n",
      "annotation summary:\n",
      "\ttRNAs: 0\n",
      "\ttmRNAs: 0\n",
      "\trRNAs: 0\n",
      "\tncRNAs: 0\n",
      "\tncRNA regions: 0\n",
      "\tCRISPR arrays: 0\n",
      "\tCDSs: 5040\n",
      "\t\thypotheticals: 218\n",
      "\t\tpseudogenes: 0\n",
      "\tsORFs: 0\n",
      "\tgaps: 0\n",
      "\toriCs/oriVs: 1\n",
      "\toriTs: 0\n",
      "\n",
      "Export annotation results to: /home/snt/SerraPHIM/data/bakta_results/CP100764.1_Serratia_plymuthica_strain_SASK2000_chromosome\n",
      "\thuman readable TSV...\n",
      "\tGFF3...\n",
      "\tINSDC GenBank & EMBL...\n",
      "\tgenome sequences...\n",
      "\tfeature nucleotide sequences...\n",
      "\ttranslated CDS sequences...\n",
      "\tfeature inferences...\n",
      "\tskip generation of circular genome plot...\n",
      "\thypothetical TSV...\n",
      "\ttranslated hypothetical CDS sequences...\n",
      "\tmachine readable JSON...\n",
      "\tGenome and annotation summary...\n",
      "\n",
      "If you use these results please cite Bakta: https://doi.org/10.1099/mgen.0.000685\n",
      "Annotation successfully finished in 3:02 [mm:ss].\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing genomes:  20%|██        | 1/5 [03:05<12:22, 185.57s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Receptor genes:      Sequence Id Type    Start     Stop Strand     Locus Tag  Gene  \\\n",
      "21      contig_1  cds    24080    24739      -  EFMCOH_00022  ompA   \n",
      "24      contig_1  cds    26068    28056      +  EFMCOH_00025   NaN   \n",
      "70      contig_1  cds    77521    78864      +  EFMCOH_00071   NaN   \n",
      "89      contig_1  cds    98711   100105      +  EFMCOH_00090   NaN   \n",
      "93      contig_1  cds   102091   103785      +  EFMCOH_00094  eptB   \n",
      "...          ...  ...      ...      ...    ...           ...   ...   \n",
      "4977    contig_1  cds  5369170  5370060      -  EFMCOH_04978  brkB   \n",
      "4991    contig_1  cds  5387832  5388425      +  EFMCOH_04992  mobA   \n",
      "5019    contig_1  cds  5420906  5422276      +  EFMCOH_05019  glmU   \n",
      "5028    contig_1  cds  5430606  5431319      +  EFMCOH_05028  hisM   \n",
      "5032    contig_1  cds  5434333  5435382      -  EFMCOH_05032   NaN   \n",
      "\n",
      "                                                Product  \\\n",
      "21                              OmpA family lipoprotein   \n",
      "24                                             Lipase 1   \n",
      "70    Glycosyltransferase sugar-binding region conta...   \n",
      "89                            Outer membrane protein D1   \n",
      "93    kdo(2)-lipid A phosphoethanolamine 7''-transfe...   \n",
      "...                                                 ...   \n",
      "4977               virulence factor BrkB family protein   \n",
      "4991       molybdenum cofactor guanylyltransferase MobA   \n",
      "5019  bifunctional UDP-N-acetylglucosamine diphospho...   \n",
      "5028  Inner membrane amino-acid ABC transporter perm...   \n",
      "5032                                      Phospholipase   \n",
      "\n",
      "                                                DbXrefs  \n",
      "21    COG:COG2885, COG:M, GO:0009279, RefSeq:WP_0138...  \n",
      "24    KEGG:K12686, RefSeq:WP_254754795.1, SO:0001217...  \n",
      "70    RefSeq:WP_013810987.1, SO:0001217, UniParc:UPI...  \n",
      "89    RefSeq:WP_254754802.1, SO:0001217, UniParc:UPI...  \n",
      "93    COG:COG2194, COG:M, EC:2.7.8.42, GO:0008484, K...  \n",
      "...                                                 ...  \n",
      "4977  COG:COG1295, COG:S, GO:0005886, RefSeq:WP_0138...  \n",
      "4991  EC:2.7.7.77, GO:0003824, GO:0006777, RefSeq:WP...  \n",
      "5019  COG:COG1207, COG:M, EC:2.3.1.157, EC:2.7.7.23,...  \n",
      "5028  COG:COG0765, COG:E, RefSeq:WP_013815074.1, SO:...  \n",
      "5032  RefSeq:WP_013815078.1, SO:0001217, UniParc:UPI...  \n",
      "\n",
      "[394 rows x 9 columns]\n",
      "Command: bakta --db /home/snt/PhageHostLearn-main/bakta_db/db --output /home/snt/SerraPHIM/data/bakta_results/OX291654.1_Serratia_marcescens_strain_SJC1039_genome_assembly --force --threads 16 --skip-trna --skip-tmrna --skip-rrna --skip-ncrna --skip-ncrna-region --skip-crispr --skip-pseudo --skip-sorf --skip-gap --skip-plot /home/snt/SerraPHIM/data/bacteria_genomes/sample/OX291654.1_Serratia_marcescens_strain_SJC1039_genome_assembly.fasta\n",
      "Parse genome sequences...\n",
      "\timported: 1\n",
      "\tfiltered & revised: 1\n",
      "\tchromosomes: 1\n",
      "\n",
      "Start annotation...\n",
      "skip tRNA prediction...\n",
      "skip tmRNA prediction...\n",
      "skip rRNA prediction...\n",
      "skip ncRNA prediction...\n",
      "skip ncRNA region prediction...\n",
      "skip CRISPR array prediction...\n",
      "predict & annotate CDSs...\n",
      "\tpredicted: 4566 \n",
      "\tdiscarded spurious: 2\n",
      "\trevised translational exceptions: 0\n",
      "\tdetected IPSs: 4332\n",
      "\tfound PSCs: 199\n",
      "\tfound PSCCs: 16\n",
      "\tlookup annotations...\n",
      "\tconduct expert systems...\n",
      "\t\tamrfinder: 12\n",
      "\t\tprotein sequences: 65\n",
      "\tcombine annotations and mark hypotheticals...\n",
      "\tanalyze hypothetical proteins: 103\n",
      "\t\tdetected Pfam hits: 3 \n",
      "\t\tcalculated proteins statistics\n",
      "\trevise special cases...\n",
      "skip sORF prediction...\n",
      "skip gap annotation...\n",
      "detect oriCs/oriVs...\n",
      "\tfound: 2\n",
      "detect oriTs...\n",
      "\tfound: 0\n",
      "apply feature overlap filters...\n",
      "select features and create locus tags...\n",
      "\tselected: 4566\n",
      "improve annotations...\n",
      "\trevised gene symbols: 66\n",
      "\n",
      "Genome statistics:\n",
      "\tGenome size: 5,048,080 bp\n",
      "\tContigs/replicons: 1\n",
      "\tGC: 60.1 %\n",
      "\tN50: 5,048,080\n",
      "\tN90: 5,048,080\n",
      "\tN ratio: 0.0 %\n",
      "\tcoding density: 87.6 %\n",
      "\n",
      "annotation summary:\n",
      "\ttRNAs: 0\n",
      "\ttmRNAs: 0\n",
      "\trRNAs: 0\n",
      "\tncRNAs: 0\n",
      "\tncRNA regions: 0\n",
      "\tCRISPR arrays: 0\n",
      "\tCDSs: 4564\n",
      "\t\thypotheticals: 103\n",
      "\t\tpseudogenes: 0\n",
      "\tsORFs: 0\n",
      "\tgaps: 0\n",
      "\toriCs/oriVs: 2\n",
      "\toriTs: 0\n",
      "\n",
      "Export annotation results to: /home/snt/SerraPHIM/data/bakta_results/OX291654.1_Serratia_marcescens_strain_SJC1039_genome_assembly\n",
      "\thuman readable TSV...\n",
      "\tGFF3...\n",
      "\tINSDC GenBank & EMBL...\n",
      "\tgenome sequences...\n",
      "\tfeature nucleotide sequences...\n",
      "\ttranslated CDS sequences...\n",
      "\tfeature inferences...\n",
      "\tskip generation of circular genome plot...\n",
      "\thypothetical TSV...\n",
      "\ttranslated hypothetical CDS sequences...\n",
      "\tmachine readable JSON...\n",
      "\tGenome and annotation summary...\n",
      "\n",
      "If you use these results please cite Bakta: https://doi.org/10.1099/mgen.0.000685\n",
      "Annotation successfully finished in 3:01 [mm:ss].\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing genomes:  40%|████      | 2/5 [06:09<09:13, 184.50s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Receptor genes:      Sequence Id Type    Start     Stop Strand     Locus Tag  Gene  \\\n",
      "8       contig_1  cds     6425     6943      +  BLNDHM_00009  ompX   \n",
      "23      contig_1  cds    18633    19304      +  BLNDHM_00024  ompR   \n",
      "27      contig_1  cds    23306    23893      +  BLNDHM_00028  ecpA   \n",
      "29      contig_1  cds    24690    27209      +  BLNDHM_00030  ecpC   \n",
      "33      contig_1  cds    30978    31691      -  BLNDHM_00034  ompR   \n",
      "...          ...  ...      ...      ...    ...           ...   ...   \n",
      "4465    contig_1  cds  4948289  4949188      +  BLNDHM_04464  rhaT   \n",
      "4473    contig_1  cds  4954569  4955246      -  BLNDHM_04472  csgG   \n",
      "4536    contig_1  cds  5015468  5016565      -  BLNDHM_04535  ompC   \n",
      "4550    contig_1  cds  5030601  5031230      +  BLNDHM_04549  paiB   \n",
      "4561    contig_1  cds  5041899  5043203      +  BLNDHM_04560  bdlA   \n",
      "\n",
      "                                                Product  \\\n",
      "8                           outer membrane protein OmpX   \n",
      "23           DNA-binding heavy metal response regulator   \n",
      "27           Common pilus major fimbrillin subunit EcpA   \n",
      "29                    Outer membrane usher protein EcpC   \n",
      "33      DNA-binding dual transcriptional regulator OmpR   \n",
      "...                                                 ...   \n",
      "4465         putative amino-acid metabolite efflux pump   \n",
      "4473  Curli production assembly/transport component ...   \n",
      "4536                                              Porin   \n",
      "4550    Protease synthase and sporulation protein PAI 2   \n",
      "4561                          Chemotaxis regulator BdlA   \n",
      "\n",
      "                                                DbXrefs  \n",
      "8     KEGG:K11934, RefSeq:WP_033637662.1, SO:0001217...  \n",
      "23    COG:COG0745, COG:TK, KEGG:K07665, RefSeq:WP_03...  \n",
      "27    KEGG:K21964, RefSeq:WP_033646847.1, SO:0001217...  \n",
      "29    KEGG:K21966, RefSeq:WP_033646843.1, SO:0001217...  \n",
      "33    COG:COG0745, COG:TK, RefSeq:WP_041034140.1, SO...  \n",
      "...                                                 ...  \n",
      "4465  COG:COG0697, COG:GER, KEGG:K15268, SO:0001217,...  \n",
      "4473  COG:COG1462, COG:W, SO:0001217, UniParc:UPI000...  \n",
      "4536  COG:COG3203, COG:M, KEGG:K11929, RefSeq:WP_033...  \n",
      "4550  COG:COG2808, COG:T, KEGG:K07734, RefSeq:WP_033...  \n",
      "4561  COG:COG0840, COG:T, RefSeq:WP_060426525.1, SO:...  \n",
      "\n",
      "[406 rows x 9 columns]\n",
      "Command: bakta --db /home/snt/PhageHostLearn-main/bakta_db/db --output /home/snt/SerraPHIM/data/bakta_results/AP013063.1_Serratia_marcescens_SM39_DNA_+_plasmids --force --threads 16 --skip-trna --skip-tmrna --skip-rrna --skip-ncrna --skip-ncrna-region --skip-crispr --skip-pseudo --skip-sorf --skip-gap --skip-plot /home/snt/SerraPHIM/data/bacteria_genomes/sample/AP013063.1_Serratia_marcescens_SM39_DNA_+_plasmids.fasta\n",
      "Parse genome sequences...\n",
      "\timported: 3\n",
      "\tfiltered & revised: 3\n",
      "\tchromosomes: 1\n",
      "\tplasmids: 2\n",
      "\n",
      "Start annotation...\n",
      "skip tRNA prediction...\n",
      "skip tmRNA prediction...\n",
      "skip rRNA prediction...\n",
      "skip ncRNA prediction...\n",
      "skip ncRNA region prediction...\n",
      "skip CRISPR array prediction...\n",
      "predict & annotate CDSs...\n",
      "\tpredicted: 4889 \n",
      "\tdiscarded spurious: 1\n",
      "\trevised translational exceptions: 0\n",
      "\tdetected IPSs: 4691\n",
      "\tfound PSCs: 161\n",
      "\tfound PSCCs: 20\n",
      "\tlookup annotations...\n",
      "\tconduct expert systems...\n",
      "\t\tamrfinder: 23\n",
      "\t\tprotein sequences: 67\n",
      "\tcombine annotations and mark hypotheticals...\n",
      "\tanalyze hypothetical proteins: 147\n",
      "\t\tdetected Pfam hits: 5 \n",
      "\t\tcalculated proteins statistics\n",
      "\trevise special cases...\n",
      "skip sORF prediction...\n",
      "skip gap annotation...\n",
      "detect oriCs/oriVs...\n",
      "\tfound: 2\n",
      "detect oriTs...\n",
      "\tfound: 0\n",
      "apply feature overlap filters...\n",
      "select features and create locus tags...\n",
      "\tselected: 4890\n",
      "improve annotations...\n",
      "\trevised gene symbols: 64\n",
      "\n",
      "Genome statistics:\n",
      "\tGenome size: 5,326,023 bp\n",
      "\tContigs/replicons: 3\n",
      "\tGC: 59.7 %\n",
      "\tN50: 5,225,577\n",
      "\tN90: 5,225,577\n",
      "\tN ratio: 0.0 %\n",
      "\tcoding density: 87.4 %\n",
      "\n",
      "annotation summary:\n",
      "\ttRNAs: 0\n",
      "\ttmRNAs: 0\n",
      "\trRNAs: 0\n",
      "\tncRNAs: 0\n",
      "\tncRNA regions: 0\n",
      "\tCRISPR arrays: 0\n",
      "\tCDSs: 4888\n",
      "\t\thypotheticals: 147\n",
      "\t\tpseudogenes: 0\n",
      "\tsORFs: 0\n",
      "\tgaps: 0\n",
      "\toriCs/oriVs: 2\n",
      "\toriTs: 0\n",
      "\n",
      "Export annotation results to: /home/snt/SerraPHIM/data/bakta_results/AP013063.1_Serratia_marcescens_SM39_DNA_+_plasmids\n",
      "\thuman readable TSV...\n",
      "\tGFF3...\n",
      "\tINSDC GenBank & EMBL...\n",
      "\tgenome sequences...\n",
      "\tfeature nucleotide sequences...\n",
      "\ttranslated CDS sequences...\n",
      "\tfeature inferences...\n",
      "\tskip generation of circular genome plot...\n",
      "\thypothetical TSV...\n",
      "\ttranslated hypothetical CDS sequences...\n",
      "\tmachine readable JSON...\n",
      "\tGenome and annotation summary...\n",
      "\n",
      "If you use these results please cite Bakta: https://doi.org/10.1099/mgen.0.000685\n",
      "Annotation successfully finished in 2:57 [mm:ss].\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing genomes:  60%|██████    | 3/5 [09:09<06:05, 182.52s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Receptor genes:      Sequence Id Type  Start   Stop Strand     Locus Tag  Gene  \\\n",
      "39      contig_1  cds  46272  48626      -  LCBGCM_00040  lptD   \n",
      "68      contig_1  cds  85151  86638      +  LCBGCM_00069  murE   \n",
      "69      contig_1  cds  86635  87996      +  LCBGCM_00070  murF   \n",
      "71      contig_1  cds  89075  90394      +  LCBGCM_00072  murD   \n",
      "74      contig_1  cds  92746  94221      +  LCBGCM_00075  murC   \n",
      "...          ...  ...    ...    ...    ...           ...   ...   \n",
      "4862    contig_3  cds  24527  24880      +  LCBGCM_04861  traA   \n",
      "4872    contig_3  cds  33715  34584      +  LCBGCM_04871  traU   \n",
      "4873    contig_3  cds  34590  34985      +  LCBGCM_04872  trbC   \n",
      "4875    contig_3  cds  36950  37720      +  LCBGCM_04874  traF   \n",
      "4885    contig_3  cds  53145  53660      +  LCBGCM_04884   NaN   \n",
      "\n",
      "                                                Product  \\\n",
      "39                            LPS assembly protein LptD   \n",
      "68    UDP-N-acetylmuramoyl-L-alanyl-D-glutamate--2,6...   \n",
      "69    UDP-N-acetylmuramoyl-tripeptide--D-alanyl-D-al...   \n",
      "71    UDP-N-acetylmuramoyl-L-alanine--D-glutamate li...   \n",
      "74               UDP-N-acetylmuramate--L-alanine ligase   \n",
      "...                                                 ...   \n",
      "4862     type IV conjugative transfer system pilin TraA   \n",
      "4872      conjugal transfer pilus assembly protein TraU   \n",
      "4873  type-F conjugative transfer system pilin assem...   \n",
      "4875  type-F conjugative transfer system pilin assem...   \n",
      "4885                     phospholipase D family protein   \n",
      "\n",
      "                                                DbXrefs  \n",
      "39    COG:COG1452, COG:M, GO:0015920, GO:0019867, GO...  \n",
      "68    COG:COG0769, COG:M, EC:6.3.2.13, GO:0005524, G...  \n",
      "69    COG:COG0770, COG:M, EC:6.3.2.10, GO:0005524, G...  \n",
      "71    COG:COG0771, COG:M, EC:6.3.2.9, GO:0005524, GO...  \n",
      "74    COG:COG0773, COG:M, EC:6.3.2.8, GO:0005524, GO...  \n",
      "...                                                 ...  \n",
      "4862  RefSeq:WP_080335488.1, SO:0001217, UniParc:UPI...  \n",
      "4872  KEGG:K12060, SO:0001217, UniRef:UniRef50_P1847...  \n",
      "4873  SO:0001217, UniRef:UniRef50_A0A6P1Q6I2, UniRef...  \n",
      "4875  KEGG:K12057, SO:0001217, UniRef:UniRef50_A0A6P...  \n",
      "4885  RefSeq:WP_041038658.1, SO:0001217, UniParc:UPI...  \n",
      "\n",
      "[416 rows x 9 columns]\n",
      "Command: bakta --db /home/snt/PhageHostLearn-main/bakta_db/db --output /home/snt/SerraPHIM/data/bakta_results/AP028466.1_Serratia_marcescens_12BL1_DNA --force --threads 16 --skip-trna --skip-tmrna --skip-rrna --skip-ncrna --skip-ncrna-region --skip-crispr --skip-pseudo --skip-sorf --skip-gap --skip-plot /home/snt/SerraPHIM/data/bacteria_genomes/sample/AP028466.1_Serratia_marcescens_12BL1_DNA.fasta\n",
      "Parse genome sequences...\n",
      "\timported: 1\n",
      "\tfiltered & revised: 1\n",
      "\tchromosomes: 1\n",
      "\n",
      "Start annotation...\n",
      "skip tRNA prediction...\n",
      "skip tmRNA prediction...\n",
      "skip rRNA prediction...\n",
      "skip ncRNA prediction...\n",
      "skip ncRNA region prediction...\n",
      "skip CRISPR array prediction...\n",
      "predict & annotate CDSs...\n",
      "\tpredicted: 4999 \n",
      "\tdiscarded spurious: 0\n",
      "\trevised translational exceptions: 0\n",
      "\tdetected IPSs: 4717\n",
      "\tfound PSCs: 228\n",
      "\tfound PSCCs: 31\n",
      "\tlookup annotations...\n",
      "\tconduct expert systems...\n",
      "\t\tamrfinder: 15\n",
      "\t\tprotein sequences: 65\n",
      "\tcombine annotations and mark hypotheticals...\n",
      "\tanalyze hypothetical proteins: 164\n",
      "\t\tdetected Pfam hits: 4 \n",
      "\t\tcalculated proteins statistics\n",
      "\trevise special cases...\n",
      "skip sORF prediction...\n",
      "skip gap annotation...\n",
      "detect oriCs/oriVs...\n",
      "\tfound: 2\n",
      "detect oriTs...\n",
      "\tfound: 0\n",
      "apply feature overlap filters...\n",
      "select features and create locus tags...\n",
      "\tselected: 5001\n",
      "improve annotations...\n",
      "\trevised gene symbols: 65\n",
      "\n",
      "Genome statistics:\n",
      "\tGenome size: 5,437,989 bp\n",
      "\tContigs/replicons: 1\n",
      "\tGC: 59.7 %\n",
      "\tN50: 5,437,989\n",
      "\tN90: 5,437,989\n",
      "\tN ratio: 0.0 %\n",
      "\tcoding density: 87.7 %\n",
      "\n",
      "annotation summary:\n",
      "\ttRNAs: 0\n",
      "\ttmRNAs: 0\n",
      "\trRNAs: 0\n",
      "\tncRNAs: 0\n",
      "\tncRNA regions: 0\n",
      "\tCRISPR arrays: 0\n",
      "\tCDSs: 4999\n",
      "\t\thypotheticals: 164\n",
      "\t\tpseudogenes: 0\n",
      "\tsORFs: 0\n",
      "\tgaps: 0\n",
      "\toriCs/oriVs: 2\n",
      "\toriTs: 0\n",
      "\n",
      "Export annotation results to: /home/snt/SerraPHIM/data/bakta_results/AP028466.1_Serratia_marcescens_12BL1_DNA\n",
      "\thuman readable TSV...\n",
      "\tGFF3...\n",
      "\tINSDC GenBank & EMBL...\n",
      "\tgenome sequences...\n",
      "\tfeature nucleotide sequences...\n",
      "\ttranslated CDS sequences...\n",
      "\tfeature inferences...\n",
      "\tskip generation of circular genome plot...\n",
      "\thypothetical TSV...\n",
      "\ttranslated hypothetical CDS sequences...\n",
      "\tmachine readable JSON...\n",
      "\tGenome and annotation summary...\n",
      "\n",
      "If you use these results please cite Bakta: https://doi.org/10.1099/mgen.0.000685\n",
      "Annotation successfully finished in 3:25 [mm:ss].\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing genomes:  80%|████████  | 4/5 [12:37<03:12, 192.56s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Receptor genes:      Sequence Id Type    Start     Stop Strand     Locus Tag  Gene  \\\n",
      "39      contig_1  cds    46257    48617      -  OANAOL_00040  lptD   \n",
      "68      contig_1  cds    85142    86629      +  OANAOL_00069  murE   \n",
      "69      contig_1  cds    86626    87987      +  OANAOL_00070  murF   \n",
      "71      contig_1  cds    89066    90385      +  OANAOL_00072  murD   \n",
      "74      contig_1  cds    92737    94212      +  OANAOL_00075  murC   \n",
      "...          ...  ...      ...      ...    ...           ...   ...   \n",
      "4955    contig_1  cds  5391145  5392347      -  OANAOL_04954  araJ   \n",
      "4983    contig_1  cds  5419174  5420607      +  OANAOL_04982   NaN   \n",
      "4987    contig_1  cds  5423699  5424667      -  OANAOL_04986   NaN   \n",
      "4990    contig_1  cds  5427576  5428073      -  OANAOL_04989  ompA   \n",
      "4997    contig_1  cds  5434195  5435064      -  OANAOL_04996  robA   \n",
      "\n",
      "                                                Product  \\\n",
      "39                            LPS assembly protein LptD   \n",
      "68    UDP-N-acetylmuramoyl-L-alanyl-D-glutamate--2,6...   \n",
      "69    UDP-N-acetylmuramoyl-tripeptide--D-alanyl-D-al...   \n",
      "71    UDP-N-acetylmuramoyl-L-alanine--D-glutamate li...   \n",
      "74               UDP-N-acetylmuramate--L-alanine ligase   \n",
      "...                                                 ...   \n",
      "4955             Purine ribonucleoside efflux pump nepI   \n",
      "4983  Glycosyltransferase RgtA/B/C/D-like domain-con...   \n",
      "4987  Rhamnolipids biosynthesis 3-oxoacyl-[acyl-carr...   \n",
      "4990                                OmpA family protein   \n",
      "4997  MDR efflux pump AcrAB transcriptional activato...   \n",
      "\n",
      "                                                DbXrefs  \n",
      "39    COG:COG1452, COG:M, GO:0015920, GO:0019867, GO...  \n",
      "68    COG:COG0769, COG:M, EC:6.3.2.13, GO:0005524, G...  \n",
      "69    COG:COG0770, COG:M, EC:6.3.2.10, GO:0005524, G...  \n",
      "71    COG:COG0771, COG:M, EC:6.3.2.9, GO:0005524, GO...  \n",
      "74    COG:COG0773, COG:M, EC:6.3.2.8, GO:0005524, GO...  \n",
      "...                                                 ...  \n",
      "4955  COG:COG2814, COG:G, RefSeq:WP_060422552.1, SO:...  \n",
      "4983  RefSeq:WP_060422519.1, SO:0001217, UniParc:UPI...  \n",
      "4987  RefSeq:WP_060422508.1, SO:0001217, UniParc:UPI...  \n",
      "4990  COG:COG2885, COG:M, RefSeq:WP_033649136.1, SO:...  \n",
      "4997  COG:COG3708, COG:K, KEGG:K05804, RefSeq:WP_033...  \n",
      "\n",
      "[425 rows x 9 columns]\n",
      "Command: bakta --db /home/snt/PhageHostLearn-main/bakta_db/db --output /home/snt/SerraPHIM/data/bakta_results/CP006250.1_Serratia_plymuthica_4Rx13_+_plasmids --force --threads 16 --skip-trna --skip-tmrna --skip-rrna --skip-ncrna --skip-ncrna-region --skip-crispr --skip-pseudo --skip-sorf --skip-gap --skip-plot /home/snt/SerraPHIM/data/bacteria_genomes/sample/CP006250.1_Serratia_plymuthica_4Rx13_+_plasmids.fasta\n",
      "Parse genome sequences...\n",
      "\timported: 2\n",
      "\tfiltered & revised: 2\n",
      "\tchromosomes: 1\n",
      "\tplasmids: 1\n",
      "\n",
      "Start annotation...\n",
      "skip tRNA prediction...\n",
      "skip tmRNA prediction...\n",
      "skip rRNA prediction...\n",
      "skip ncRNA prediction...\n",
      "skip ncRNA region prediction...\n",
      "skip CRISPR array prediction...\n",
      "predict & annotate CDSs...\n",
      "\tpredicted: 4918 \n",
      "\tdiscarded spurious: 1\n",
      "\trevised translational exceptions: 0\n",
      "\tdetected IPSs: 4554\n",
      "\tfound PSCs: 242\n",
      "\tfound PSCCs: 47\n",
      "\tlookup annotations...\n",
      "\tconduct expert systems...\n",
      "\t\tamrfinder: 7\n",
      "\t\tprotein sequences: 51\n",
      "\tcombine annotations and mark hypotheticals...\n",
      "\tanalyze hypothetical proteins: 206\n",
      "\t\tdetected Pfam hits: 11 \n",
      "\t\tcalculated proteins statistics\n",
      "\trevise special cases...\n",
      "skip sORF prediction...\n",
      "skip gap annotation...\n",
      "detect oriCs/oriVs...\n",
      "\tfound: 1\n",
      "detect oriTs...\n",
      "\tfound: 0\n",
      "apply feature overlap filters...\n",
      "select features and create locus tags...\n",
      "\tselected: 4918\n",
      "improve annotations...\n",
      "\trevised gene symbols: 35\n",
      "\n",
      "Genome statistics:\n",
      "\tGenome size: 5,403,731 bp\n",
      "\tContigs/replicons: 2\n",
      "\tGC: 56.1 %\n",
      "\tN50: 5,328,010\n",
      "\tN90: 5,328,010\n",
      "\tN ratio: 0.0 %\n",
      "\tcoding density: 87.0 %\n",
      "\n",
      "annotation summary:\n",
      "\ttRNAs: 0\n",
      "\ttmRNAs: 0\n",
      "\trRNAs: 0\n",
      "\tncRNAs: 0\n",
      "\tncRNA regions: 0\n",
      "\tCRISPR arrays: 0\n",
      "\tCDSs: 4917\n",
      "\t\thypotheticals: 206\n",
      "\t\tpseudogenes: 0\n",
      "\tsORFs: 0\n",
      "\tgaps: 0\n",
      "\toriCs/oriVs: 1\n",
      "\toriTs: 0\n",
      "\n",
      "Export annotation results to: /home/snt/SerraPHIM/data/bakta_results/CP006250.1_Serratia_plymuthica_4Rx13_+_plasmids\n",
      "\thuman readable TSV...\n",
      "\tGFF3...\n",
      "\tINSDC GenBank & EMBL...\n",
      "\tgenome sequences...\n",
      "\tfeature nucleotide sequences...\n",
      "\ttranslated CDS sequences...\n",
      "\tfeature inferences...\n",
      "\tskip generation of circular genome plot...\n",
      "\thypothetical TSV...\n",
      "\ttranslated hypothetical CDS sequences...\n",
      "\tmachine readable JSON...\n",
      "\tGenome and annotation summary...\n",
      "\n",
      "If you use these results please cite Bakta: https://doi.org/10.1099/mgen.0.000685\n",
      "Annotation successfully finished in 3:16 [mm:ss].\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing genomes: 100%|██████████| 5/5 [15:56<00:00, 191.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Receptor genes:      Sequence Id Type  Start    Stop Strand     Locus Tag  Gene  \\\n",
      "21      contig_1  cds  23947   24606      -  EPGMHD_00022  ompA   \n",
      "24      contig_1  cds  25935   27923      +  EPGMHD_00025   NaN   \n",
      "67      contig_1  cds  76026   76334      +  EPGMHD_00068  rhuM   \n",
      "85      contig_1  cds  96510   97904      +  EPGMHD_00086   NaN   \n",
      "89      contig_1  cds  99891  101585      +  EPGMHD_00090  eptB   \n",
      "...          ...  ...    ...     ...    ...           ...   ...   \n",
      "4879    contig_2  cds  31873   32868      +  EPGMHD_04879  traU   \n",
      "4880    contig_2  cds  32880   33485      +  EPGMHD_04880  trbC   \n",
      "4882    contig_2  cds  35503   36288      +  EPGMHD_04882  traF   \n",
      "4887    contig_2  cds  38412   39791      +  EPGMHD_04887  traH   \n",
      "4893    contig_2  cds  52066   52605      +  EPGMHD_04893   NaN   \n",
      "\n",
      "                                                Product  \\\n",
      "21                              OmpA family lipoprotein   \n",
      "24                                             Lipase 1   \n",
      "67                               virulence protein RhuM   \n",
      "85                            Outer membrane protein D1   \n",
      "89    kdo(2)-lipid A phosphoethanolamine 7''-transfe...   \n",
      "...                                                 ...   \n",
      "4879      conjugal transfer pilus assembly protein TraU   \n",
      "4880  type-F conjugative transfer system pilin assem...   \n",
      "4882  type-F conjugative transfer system pilin assem...   \n",
      "4887      conjugal transfer pilus assembly protein TraH   \n",
      "4893                                    phospholipase D   \n",
      "\n",
      "                                                DbXrefs  \n",
      "21    COG:COG2885, COG:M, GO:0009279, RefSeq:WP_0049...  \n",
      "24    KEGG:K12686, RefSeq:WP_004950619.1, SO:0001217...  \n",
      "67    SO:0001217, UniRef:UniRef50_A0A1B8YK09, UniRef...  \n",
      "85    RefSeq:WP_004950790.1, SO:0001217, UniParc:UPI...  \n",
      "89    COG:COG2194, COG:M, EC:2.7.8.42, GO:0008484, K...  \n",
      "...                                                 ...  \n",
      "4879  KEGG:K12060, RefSeq:WP_004953915.1, SO:0001217...  \n",
      "4880  RefSeq:WP_006322777.1, SO:0001217, UniParc:UPI...  \n",
      "4882  KEGG:K12057, RefSeq:WP_004953823.1, SO:0001217...  \n",
      "4887  KEGG:K12072, RefSeq:WP_020453980.1, SO:0001217...  \n",
      "4893  RefSeq:WP_230554984.1, SO:0001217, UniParc:UPI...  \n",
      "\n",
      "[398 rows x 9 columns]\n",
      "Receptor extraction complete. Results saved to /home/snt/SerraPHIM/data/bakta_results/Bact_receptors_training.json and /home/snt/SerraPHIM/data/bakta_results/Bact_receptors_verbose_training.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Run Bakta\n",
    "product_keywords = [\n",
    "    \"capsule\", \"polysaccharide\", \"-polysaccharide\", \"LPS\",\n",
    "    \"oligosaccharide polymerase\",\n",
    "    \"UDP-\", \"GDP-\",\n",
    "    \"mannose\",\n",
    "    \"guanylyltransferase\",\n",
    "    \"glycosyltransferase\",\n",
    "    \"flippase\",\n",
    "    \"-antigen\", \"O-antigen\",\n",
    "    \"outer membrane\", \"outer core\", \"omp-\",\n",
    "    \"mannosyltransferase\",\n",
    "    \"fimbria-\",\n",
    "    \"pili-\", \"pilus\",\n",
    "    \"adhesin\",\n",
    "    \"cellulose\",\n",
    "    \"curli\",\n",
    "    # \"receptor\", # too generic? any receptors not related to surface receptors? \n",
    "    \"siderophore receptor\", \"siderophore\",\n",
    "    \"enterobactin receptor\",\n",
    "    \"aerobactin receptor\",\n",
    "    \"yersiniabactin receptor\",\n",
    "    \"salmochelin receptor\",\n",
    "    \"phage receptor\",\n",
    "    # \"prophage\", # how is a possible prophage existence affecting the susceptibility to other phages?\n",
    "    # \"phage\", \"tail fiber\", # + other phage components... (CDS in prophage regions?)\n",
    "    \"porin\", \"-porin\",\n",
    "    \"efflux pump\",\n",
    "    \"hemagglutinin\", \"haemagglutinin\",\n",
    "    \"lectin\",\n",
    "    \"lipid\", \"-lipid\", \n",
    "    # \"lipoprotein\", # too generic? any lipoproteins not related to cell wall / surface receptors?\n",
    "    \"beta-barrel assembly\",\n",
    "    \"autotransporter\",\n",
    "    \"pathway protein\",\n",
    "    \"flagell-\",\n",
    "    \"chemotaxis\",\n",
    "    \"transporter membrane\", \"transmembrane\",\n",
    "    \"TonB-dependent receptor\", \"TonB\", \"TonB-\",\n",
    "    \"chitinase\", \"lecithinase\", \"hemolysin\", \n",
    "    \"lipase\", \"-lipase\", \n",
    "    \"protease\", \"-protease\",\n",
    "    \"exoenzyme-\", \"virulence\", \"invasion\"\n",
    "]\n",
    "gene_keywords = [\n",
    "    \"wza\", # (capsule export)\n",
    "    \"wzb\", # (tyrosine-protein phosphatase)\n",
    "    \"wzc\", # (capsule synthesis tyrosine kinase)\n",
    "    \"galF\", # (UDP-glucose pyrophosphorylase)\n",
    "    \"ugd\", # (UDP-glucose dehydrogenase)\n",
    "    \"manC\", # (GDP-mannose pyrophosphorylase)\n",
    "    \"manB\", # (mannose-1-phosphate guanylyltransferase)\n",
    "    \"Wzx\", # (O-antigen flippase)\n",
    "    \"Wzy\", # (O-antigen polymerase)\n",
    "    \"WaaL\", # (O-antigen ligase)\n",
    "    \"Waa-\", # (core LPS biosynthesis)\n",
    "    #\"magA\", # (K1-specific capsular serotype proteins)\n",
    "    #\"rmpA\", # (K2-specific capsular serotype proteins)\n",
    "    \"OmpA\", \"OmpX\", \"Omp-\", # (outer membrane proteins)\n",
    "    #\"OmpK-\", # (Klebsiella-specific outer membrane proteins)\n",
    "    \"TolC\", # (outer membrane proteins)\n",
    "    \"OprD\", \"Opr-\", # (outer membrane proteins)\n",
    "    \"HasR\", \"HasB\", # (TonB-dependent outer membrane receptor complex)\n",
    "    \"lpx-\", # (lipid A biosynthesis: lpxA, lpxC, lpxD)\n",
    "    \"WbbP\", # (mannosyltransferase)\n",
    "    \"fimA\", \"fimH\", # (Type 1 fimbriae)\n",
    "    \"mrkA\", \"mrkB\", \"mrkC\", # (Type 3 fimbriae)\n",
    "    \"PilQ\", \"PilX\", # (Adhesins)\n",
    "    \"csgA\", \"csgB\", # (Curli fibers)\n",
    "    \"FliD\", # (Flagellar hook-associated proteins)\n",
    "    \"CheA\", \"CheB\", \"CheR\", \"CheY\", \"CheW\", \"CheZ\", \"Che-\", # (Chemotaxis proteins)\n",
    "    \"FepA\", \"Fiu\", # (Enterobactin receptors)\n",
    "    \"IutA\", # (Aerobactin receptors)\n",
    "    \"FyuA\", # (Yersiniabactin receptors)\n",
    "    \"IroN\" # (Salmochelin receptors)\n",
    "]\n",
    "sp.run_bakta_and_extract_receptors(\n",
    "    bact_genomes_sample_path, # bact_genomes_path, \n",
    "    bakta_results_path, \n",
    "    bakta_db_path, \n",
    "    gene_keywords, product_keywords,\n",
    "    data_suffix,\n",
    "    threads=16,  # 128 cores in yakuza server\n",
    "    training=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c28dc9db",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Embedding sequences: 100%|██████████| 14/14 [01:00<00:00,  4.32s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved embeddings to /home/snt/SerraPHIM/data/esm2_embeddings_rbp_training.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# 3 - FEATURE CONSTRUCTION\n",
    "# --------------------------------------------------\n",
    "\n",
    "# ESM-2 features for RBPs\n",
    "sf.compute_esm2_embeddings_rbp_improved(\n",
    "    data_path, \n",
    "    data_suffix, \n",
    "    batch_size=16\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ae3b0721",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Will process CP100764.1_Serratia_plymuthica_strain_SASK2000_chromosome_288 solo (length=1880 AA)\n",
      "Will process CP100764.1_Serratia_plymuthica_strain_SASK2000_chromosome_320 solo (length=2154 AA)\n",
      "Will process CP100764.1_Serratia_plymuthica_strain_SASK2000_chromosome_342 solo (length=3314 AA)\n",
      "Will process CP100764.1_Serratia_plymuthica_strain_SASK2000_chromosome_349 solo (length=1607 AA)\n",
      "Will process OX291654.1_Serratia_marcescens_strain_SJC1039_genome_assembly_92 solo (length=3548 AA)\n",
      "Will process OX291654.1_Serratia_marcescens_strain_SJC1039_genome_assembly_275 solo (length=1608 AA)\n",
      "Will process AP013063.1_Serratia_marcescens_SM39_DNA_+_plasmids_141 solo (length=3642 AA)\n",
      "Will process AP013063.1_Serratia_marcescens_SM39_DNA_+_plasmids_324 solo (length=1608 AA)\n",
      "Will process AP028466.1_Serratia_marcescens_12BL1_DNA_147 solo (length=2722 AA)\n",
      "Will process AP028466.1_Serratia_marcescens_12BL1_DNA_271 solo (length=2023 AA)\n",
      "Will process AP028466.1_Serratia_marcescens_12BL1_DNA_338 solo (length=1608 AA)\n",
      "Will process CP006250.1_Serratia_plymuthica_4Rx13_+_plasmids_285 solo (length=1959 AA)\n",
      "Will process CP006250.1_Serratia_plymuthica_4Rx13_+_plasmids_321 solo (length=2680 AA)\n",
      "Will process CP006250.1_Serratia_plymuthica_4Rx13_+_plasmids_351 solo (length=1608 AA)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Embedding batches: 100%|██████████| 127/127 [1:46:50<00:00, 50.48s/it]\n",
      "Embedding batches: 100%|██████████| 14/14 [03:05<00:00, 13.27s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved embeddings to: /home/snt/SerraPHIM/data/esm2_embeddings_receptors_training.csv\n"
     ]
    }
   ],
   "source": [
    "# ESM-2 features for bacterial receptors \n",
    "sf.compute_esm2_embeddings_receptors(\n",
    "    data_path, \n",
    "    bakta_results_path, \n",
    "    data_suffix,\n",
    "    aggregate_by_accession=True,  # Set to true during internship\n",
    "    batch_size=16,\n",
    "    skip_long_seq=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6b51c25c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building feature matrix: 100%|██████████| 20/20 [00:00<00:00, 1008.08it/s]\n"
     ]
    }
   ],
   "source": [
    "# Construct feature matrices\n",
    "rbp_embeddings_path = data_path + '/esm2_embeddings_rbp' + data_suffix + '.csv'\n",
    "receptors_embeddings_path = data_path + '/esm2_embeddings_receptors' + data_suffix + '.csv'\n",
    "interaction_path = data_path + '/interactions'\n",
    "\n",
    "features_esm2, labels, groups_receptors, groups_phage = sf.construct_feature_matrices_receptors(\n",
    "    interaction_path, \n",
    "    data_suffix, \n",
    "    receptors_embeddings_path, \n",
    "    rbp_embeddings_path\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9497f36d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[np.float64(-0.014332249), np.float64(-0.06588216),\n",
       "         np.float64(-0.0023698194), ..., np.float64(-0.071024075),\n",
       "         np.float64(-0.013574367666666665),\n",
       "         np.float64(0.10997433999999999)],\n",
       "        [np.float64(-0.014332249), np.float64(-0.06588216),\n",
       "         np.float64(-0.0023698194), ..., np.float64(-0.12421417),\n",
       "         np.float64(-0.0037242933), np.float64(0.1801741)],\n",
       "        [np.float64(-0.014332249), np.float64(-0.06588216),\n",
       "         np.float64(-0.0023698194), ..., np.float64(-0.07182163700000001),\n",
       "         np.float64(0.0048919260000000004), np.float64(0.0896083225)],\n",
       "        ...,\n",
       "        [np.float64(-0.015825137), np.float64(-0.06701858),\n",
       "         np.float64(-0.001265623), ..., np.float64(-0.12421417),\n",
       "         np.float64(-0.0037242933), np.float64(0.1801741)],\n",
       "        [np.float64(-0.015825137), np.float64(-0.06701858),\n",
       "         np.float64(-0.001265623), ..., np.float64(-0.07182163700000001),\n",
       "         np.float64(0.0048919260000000004), np.float64(0.0896083225)],\n",
       "        [np.float64(-0.015825137), np.float64(-0.06701858),\n",
       "         np.float64(-0.001265623), ..., np.float64(-0.07648365900000001),\n",
       "         np.float64(0.025897467374999997),\n",
       "         np.float64(0.07415039550000001)]], shape=(20, 2560), dtype=object),\n",
       " array([1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0]),\n",
       " ['AP013063.1',\n",
       "  'AP013063.1',\n",
       "  'AP013063.1',\n",
       "  'AP013063.1',\n",
       "  'AP028466.1',\n",
       "  'AP028466.1',\n",
       "  'AP028466.1',\n",
       "  'AP028466.1',\n",
       "  'CP006250.1',\n",
       "  'CP006250.1',\n",
       "  'CP006250.1',\n",
       "  'CP006250.1',\n",
       "  'CP100764.1',\n",
       "  'CP100764.1',\n",
       "  'CP100764.1',\n",
       "  'CP100764.1',\n",
       "  'OX291654.1',\n",
       "  'OX291654.1',\n",
       "  'OX291654.1',\n",
       "  'OX291654.1'],\n",
       " ['NC_021563.1',\n",
       "  'IMGVR_UViG_640753048_000002_640753048_640753089',\n",
       "  'IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631',\n",
       "  'NC_020083.1',\n",
       "  'NC_021563.1',\n",
       "  'IMGVR_UViG_640753048_000002_640753048_640753089',\n",
       "  'IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631',\n",
       "  'NC_020083.1',\n",
       "  'NC_021563.1',\n",
       "  'IMGVR_UViG_640753048_000002_640753048_640753089',\n",
       "  'IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631',\n",
       "  'NC_020083.1',\n",
       "  'NC_021563.1',\n",
       "  'IMGVR_UViG_640753048_000002_640753048_640753089',\n",
       "  'IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631',\n",
       "  'NC_020083.1',\n",
       "  'NC_021563.1',\n",
       "  'IMGVR_UViG_640753048_000002_640753048_640753089',\n",
       "  'IMGVR_UViG_3300031415_000018_3300031415_Ga0308172_1000631',\n",
       "  'NC_020083.1'])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_esm2, labels, groups_receptors, groups_phage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "999bc1a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/xgboost/training.py:183: UserWarning: [18:13:00] WARNING: /workspace/src/learner.cc:738: \n",
      "Parameters: { \"use_label_encoder\" } are not used.\n",
      "\n",
      "  bst.update(dtrain, iteration=i, fobj=obj)\n"
     ]
    }
   ],
   "source": [
    "# 4 - TRAINING & EVALUATING MODELS\n",
    "# --------------------------------------------------\n",
    "\n",
    "X_full = features_esm2\n",
    "y_full = labels\n",
    "groups_receptors = np.array(groups_receptors)\n",
    "groups_phage = np.array(groups_phage)\n",
    "\n",
    "imbalance = np.sum(y_full == 1) / np.sum(y_full == 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a415b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "search_space = {\n",
    "    \"learning_rate\": tune.loguniform(1e-3, 0.3),\n",
    "    \"max_depth\": tune.choice([3, 5, 7, 9]),\n",
    "    \"n_estimators\": tune.choice([100, 200, 300]),\n",
    "    \"scale_pos_weight\": tune.choice(1 / imbalance),\n",
    "    \"subsample\": tune.uniform(0.6, 1.0),\n",
    "    \"colsample_bytree\": tune.uniform(0.6, 1.0)\n",
    "}\n",
    "\n",
    "def train_xgb_tune(config, X, y):\n",
    "    # Stratified train/test split\n",
    "    X_train, X_val, y_train, y_val = train_test_split(\n",
    "        X, y, test_size=0.2, stratify=y, random_state=42\n",
    "    )\n",
    "\n",
    "    model = XGBClassifier(\n",
    "        learning_rate=config[\"learning_rate\"],\n",
    "        max_depth=config[\"max_depth\"],\n",
    "        n_estimators=config[\"n_estimators\"],\n",
    "        scale_pos_weight=config[\"scale_pos_weight\"],\n",
    "        subsample=config[\"subsample\"],\n",
    "        colsample_bytree=config[\"colsample_bytree\"],\n",
    "        eval_metric=\"logloss\",\n",
    "        use_label_encoder=False,\n",
    "        n_jobs=8\n",
    "    )\n",
    "    model.fit(X_train, y_train)\n",
    "    preds = model.predict_proba(X_val)[:, 1]\n",
    "    auc = roc_auc_score(y_val, preds)\n",
    "    \n",
    "    tune.report({\"auc\": auc})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "bb17b370",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-07-25 19:52:21,600\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div class=\"tuneStatus\">\n",
       "  <div style=\"display: flex;flex-direction: row\">\n",
       "    <div style=\"display: flex;flex-direction: column;\">\n",
       "      <h3>Tune Status</h3>\n",
       "      <table>\n",
       "<tbody>\n",
       "<tr><td>Current time:</td><td>2025-07-25 19:53:54</td></tr>\n",
       "<tr><td>Running for: </td><td>00:01:32.38        </td></tr>\n",
       "<tr><td>Memory:      </td><td>6.4/15.3 GiB       </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "    </div>\n",
       "    <div class=\"vDivider\"></div>\n",
       "    <div class=\"systemInfo\">\n",
       "      <h3>System Info</h3>\n",
       "      Using AsyncHyperBand: num_stopped=0<br>Bracket: Iter 20.000: None | Iter 5.000: None<br>Logical resource usage: 8.0/16 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:G)\n",
       "    </div>\n",
       "    \n",
       "  </div>\n",
       "  <div class=\"hDivider\"></div>\n",
       "  <div class=\"trialStatus\">\n",
       "    <h3>Trial Status</h3>\n",
       "    <table>\n",
       "<thead>\n",
       "<tr><th>Trial name                </th><th>status    </th><th>loc                </th><th style=\"text-align: right;\">  colsample_bytree</th><th style=\"text-align: right;\">  learning_rate</th><th style=\"text-align: right;\">  max_depth</th><th style=\"text-align: right;\">  n_estimators</th><th style=\"text-align: right;\">  scale_pos_weight</th><th style=\"text-align: right;\">  subsample</th><th style=\"text-align: right;\">  iter</th><th style=\"text-align: right;\">  total time (s)</th><th style=\"text-align: right;\">  auc</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>train_xgb_tune_1d5bc_00000</td><td>TERMINATED</td><td>172.21.5.247:389643</td><td style=\"text-align: right;\">          0.765902</td><td style=\"text-align: right;\">     0.0229313 </td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          0.778575</td><td style=\"text-align: right;\">   0.602845</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.238736</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00001</td><td>TERMINATED</td><td>172.21.5.247:389644</td><td style=\"text-align: right;\">          0.88709 </td><td style=\"text-align: right;\">     0.0293344 </td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          0.951049</td><td style=\"text-align: right;\">   0.733634</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.197464</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00002</td><td>TERMINATED</td><td>172.21.5.247:389944</td><td style=\"text-align: right;\">          0.644706</td><td style=\"text-align: right;\">     0.0364152 </td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          1.72509 </td><td style=\"text-align: right;\">   0.895299</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.122766</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00003</td><td>TERMINATED</td><td>172.21.5.247:389949</td><td style=\"text-align: right;\">          0.716024</td><td style=\"text-align: right;\">     0.0432882 </td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          0.206944</td><td style=\"text-align: right;\">   0.791245</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.144898</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00004</td><td>TERMINATED</td><td>172.21.5.247:390235</td><td style=\"text-align: right;\">          0.751918</td><td style=\"text-align: right;\">     0.00683854</td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          6.61129 </td><td style=\"text-align: right;\">   0.910807</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.17205 </td><td style=\"text-align: right;\">1    </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00005</td><td>TERMINATED</td><td>172.21.5.247:390236</td><td style=\"text-align: right;\">          0.760148</td><td style=\"text-align: right;\">     0.00428036</td><td style=\"text-align: right;\">          3</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          1.00642 </td><td style=\"text-align: right;\">   0.879841</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.208725</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00006</td><td>TERMINATED</td><td>172.21.5.247:390527</td><td style=\"text-align: right;\">          0.91163 </td><td style=\"text-align: right;\">     0.137713  </td><td style=\"text-align: right;\">          9</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          0.976634</td><td style=\"text-align: right;\">   0.746562</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.159523</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00007</td><td>TERMINATED</td><td>172.21.5.247:390528</td><td style=\"text-align: right;\">          0.766385</td><td style=\"text-align: right;\">     0.0244761 </td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          0.991594</td><td style=\"text-align: right;\">   0.728129</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.131692</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00008</td><td>TERMINATED</td><td>172.21.5.247:390816</td><td style=\"text-align: right;\">          0.644272</td><td style=\"text-align: right;\">     0.0144879 </td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          3.98733 </td><td style=\"text-align: right;\">   0.985618</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.291236</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00009</td><td>TERMINATED</td><td>172.21.5.247:390821</td><td style=\"text-align: right;\">          0.931815</td><td style=\"text-align: right;\">     0.0541606 </td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          1.0747  </td><td style=\"text-align: right;\">   0.626137</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.242155</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00010</td><td>TERMINATED</td><td>172.21.5.247:391113</td><td style=\"text-align: right;\">          0.660019</td><td style=\"text-align: right;\">     0.00134288</td><td style=\"text-align: right;\">          9</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          0.330794</td><td style=\"text-align: right;\">   0.809899</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.201341</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00011</td><td>TERMINATED</td><td>172.21.5.247:391114</td><td style=\"text-align: right;\">          0.816053</td><td style=\"text-align: right;\">     0.0148243 </td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          0.158924</td><td style=\"text-align: right;\">   0.869892</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.26443 </td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00012</td><td>TERMINATED</td><td>172.21.5.247:391413</td><td style=\"text-align: right;\">          0.823997</td><td style=\"text-align: right;\">     0.00111556</td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          0.104535</td><td style=\"text-align: right;\">   0.712391</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.191135</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00013</td><td>TERMINATED</td><td>172.21.5.247:391418</td><td style=\"text-align: right;\">          0.916416</td><td style=\"text-align: right;\">     0.00823669</td><td style=\"text-align: right;\">          9</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          6.87115 </td><td style=\"text-align: right;\">   0.969052</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.300949</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00014</td><td>TERMINATED</td><td>172.21.5.247:391709</td><td style=\"text-align: right;\">          0.713599</td><td style=\"text-align: right;\">     0.0451827 </td><td style=\"text-align: right;\">          3</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          5.50325 </td><td style=\"text-align: right;\">   0.934227</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.25098 </td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00015</td><td>TERMINATED</td><td>172.21.5.247:391701</td><td style=\"text-align: right;\">          0.696561</td><td style=\"text-align: right;\">     0.00524933</td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          0.818416</td><td style=\"text-align: right;\">   0.687027</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.270088</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00016</td><td>TERMINATED</td><td>172.21.5.247:391999</td><td style=\"text-align: right;\">          0.694752</td><td style=\"text-align: right;\">     0.221203  </td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          0.123781</td><td style=\"text-align: right;\">   0.746663</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.100479</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00017</td><td>TERMINATED</td><td>172.21.5.247:391994</td><td style=\"text-align: right;\">          0.676245</td><td style=\"text-align: right;\">     0.00216305</td><td style=\"text-align: right;\">          9</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          0.892733</td><td style=\"text-align: right;\">   0.775239</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.148965</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00018</td><td>TERMINATED</td><td>172.21.5.247:392286</td><td style=\"text-align: right;\">          0.715046</td><td style=\"text-align: right;\">     0.0719709 </td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          0.876656</td><td style=\"text-align: right;\">   0.6619  </td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.134739</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00019</td><td>TERMINATED</td><td>172.21.5.247:392291</td><td style=\"text-align: right;\">          0.969151</td><td style=\"text-align: right;\">     0.0052904 </td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          5.49979 </td><td style=\"text-align: right;\">   0.758338</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.215001</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00020</td><td>TERMINATED</td><td>172.21.5.247:392579</td><td style=\"text-align: right;\">          0.694468</td><td style=\"text-align: right;\">     0.00453632</td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          3.65702 </td><td style=\"text-align: right;\">   0.685261</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.124386</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00021</td><td>TERMINATED</td><td>172.21.5.247:392584</td><td style=\"text-align: right;\">          0.951939</td><td style=\"text-align: right;\">     0.00138001</td><td style=\"text-align: right;\">          9</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          0.155783</td><td style=\"text-align: right;\">   0.643928</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.122756</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00022</td><td>TERMINATED</td><td>172.21.5.247:392876</td><td style=\"text-align: right;\">          0.76255 </td><td style=\"text-align: right;\">     0.0273338 </td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          1.08948 </td><td style=\"text-align: right;\">   0.720805</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.194945</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00023</td><td>TERMINATED</td><td>172.21.5.247:392871</td><td style=\"text-align: right;\">          0.794262</td><td style=\"text-align: right;\">     0.00545532</td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          2.46696 </td><td style=\"text-align: right;\">   0.686231</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.219571</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00024</td><td>TERMINATED</td><td>172.21.5.247:393173</td><td style=\"text-align: right;\">          0.703512</td><td style=\"text-align: right;\">     0.0165834 </td><td style=\"text-align: right;\">          5</td><td style=\"text-align: right;\">           100</td><td style=\"text-align: right;\">          2.19032 </td><td style=\"text-align: right;\">   0.829681</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.146585</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00025</td><td>TERMINATED</td><td>172.21.5.247:393168</td><td style=\"text-align: right;\">          0.641518</td><td style=\"text-align: right;\">     0.00449413</td><td style=\"text-align: right;\">          9</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          5.15471 </td><td style=\"text-align: right;\">   0.876226</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.200309</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00026</td><td>TERMINATED</td><td>172.21.5.247:393471</td><td style=\"text-align: right;\">          0.916495</td><td style=\"text-align: right;\">     0.00864626</td><td style=\"text-align: right;\">          3</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          0.246689</td><td style=\"text-align: right;\">   0.832715</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.178739</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00027</td><td>TERMINATED</td><td>172.21.5.247:393472</td><td style=\"text-align: right;\">          0.865697</td><td style=\"text-align: right;\">     0.0255206 </td><td style=\"text-align: right;\">          7</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          0.677555</td><td style=\"text-align: right;\">   0.894879</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.173183</td><td style=\"text-align: right;\">0.875</td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00028</td><td>TERMINATED</td><td>172.21.5.247:393763</td><td style=\"text-align: right;\">          0.8035  </td><td style=\"text-align: right;\">     0.00808263</td><td style=\"text-align: right;\">          3</td><td style=\"text-align: right;\">           200</td><td style=\"text-align: right;\">          0.162437</td><td style=\"text-align: right;\">   0.79884 </td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.138553</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00029</td><td>TERMINATED</td><td>172.21.5.247:393768</td><td style=\"text-align: right;\">          0.630427</td><td style=\"text-align: right;\">     0.00101054</td><td style=\"text-align: right;\">          3</td><td style=\"text-align: right;\">           300</td><td style=\"text-align: right;\">          6.34219 </td><td style=\"text-align: right;\">   0.810598</td><td style=\"text-align: right;\">     1</td><td style=\"text-align: right;\">        0.240738</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "  </div>\n",
       "</div>\n",
       "<style>\n",
       ".tuneStatus {\n",
       "  color: var(--jp-ui-font-color1);\n",
       "}\n",
       ".tuneStatus .systemInfo {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus td {\n",
       "  white-space: nowrap;\n",
       "}\n",
       ".tuneStatus .trialStatus {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".tuneStatus .hDivider {\n",
       "  border-bottom-width: var(--jp-border-width);\n",
       "  border-bottom-color: var(--jp-border-color0);\n",
       "  border-bottom-style: solid;\n",
       "}\n",
       ".tuneStatus .vDivider {\n",
       "  border-left-width: var(--jp-border-width);\n",
       "  border-left-color: var(--jp-border-color0);\n",
       "  border-left-style: solid;\n",
       "  margin: 0.5em 1em 0.5em 1em;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div class=\"trialProgress\">\n",
       "  <h3>Trial Progress</h3>\n",
       "  <table>\n",
       "<thead>\n",
       "<tr><th>Trial name                </th><th style=\"text-align: right;\">  auc</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>train_xgb_tune_1d5bc_00000</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00001</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00002</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00003</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00004</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00005</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00006</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00007</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00008</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00009</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00010</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00011</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00012</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00013</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00014</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00015</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00016</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00017</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00018</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00019</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00020</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00021</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00022</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00023</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00024</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00025</td><td style=\"text-align: right;\">0.75 </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00026</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00027</td><td style=\"text-align: right;\">0.875</td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00028</td><td style=\"text-align: right;\">0.5  </td></tr>\n",
       "<tr><td>train_xgb_tune_1d5bc_00029</td><td style=\"text-align: right;\">1    </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>\n",
       "<style>\n",
       ".trialProgress {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  color: var(--jp-ui-font-color1);\n",
       "}\n",
       ".trialProgress h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".trialProgress td {\n",
       "  white-space: nowrap;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-07-25 19:53:54,172\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/snt/SerraPHIM/data/ray_tune_results/xgb_receptor_tune' in 0.0095s.\n",
      "2025-07-25 19:53:54,183\tINFO tune.py:1041 -- Total run time: 92.58 seconds (92.37 seconds for the tuning loop).\n"
     ]
    }
   ],
   "source": [
    "analysis = tune.run(\n",
    "    partial(train_xgb_tune, X=X_full, y=y_full),\n",
    "    config=search_space,\n",
    "    num_samples=30,  # increase for more exhaustive search\n",
    "    metric=\"auc\",\n",
    "    mode=\"max\",\n",
    "    scheduler=ASHAScheduler(max_t=50, grace_period=5),\n",
    "    resources_per_trial={\"cpu\": 8},\n",
    "    storage_path=f\"{data_path}/ray_tune_results\",\n",
    "    name=\"xgb_receptor_tune\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39f7d7b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best hyperparameters: {'learning_rate': 0.006838535213235038, 'max_depth': 5, 'n_estimators': 100, 'scale_pos_weight': np.float64(0.7777777777777777), 'subsample': 0.9108073952348492, 'colsample_bytree': 0.751917908204838}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/xgboost/training.py:183: UserWarning: [20:00:46] WARNING: /workspace/src/learner.cc:738: \n",
      "Parameters: { \"use_label_encoder\" } are not used.\n",
      "\n",
      "  bst.update(dtrain, iteration=i, fobj=obj)\n"
     ]
    }
   ],
   "source": [
    "best_config = analysis.get_best_config(metric=\"auc\", mode=\"max\")\n",
    "print(\"Best hyperparameters:\", best_config)\n",
    "\n",
    "# Train final model using best config\n",
    "best_model = XGBClassifier(\n",
    "    **best_config,\n",
    "    eval_metric='logloss',\n",
    "    use_label_encoder=False,\n",
    "    n_jobs=8\n",
    ")\n",
    "best_model.fit(X_full, y_full)\n",
    "best_model.save_model(f\"{results_path}/xgb_receptor_model_tuned{data_suffix}.json\")"
   ]
  },
  {
   "cell_type": "raw",
   "id": "20b1d672",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "\n",
    "xgb_final = XGBClassifier(\n",
    "    scale_pos_weight=1 / imbalance,\n",
    "    learning_rate=0.3,\n",
    "    n_estimators=250,\n",
    "    max_depth=7,\n",
    "    n_jobs=8,\n",
    "    eval_metric='logloss',\n",
    "    use_label_encoder=False\n",
    ")\n",
    "xgb_final.fit(X_full, y_full)\n",
    "xgb_final.save_model(f\"{data_path}/xgb_receptor_model{data_suffix}.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "382b7b7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique groups: 5\n"
     ]
    }
   ],
   "source": [
    "# GROUP STRATEGY — choose one of: 'receptor', 'phage', or 'both'\n",
    "group_by = 'receptor'\n",
    "\n",
    "if group_by == 'receptor':\n",
    "    groups = np.array(groups_receptors)\n",
    "elif group_by == 'phage':\n",
    "    groups = np.array(groups_phage)\n",
    "elif group_by == 'both':\n",
    "    # Optional: define each group as a (phage, receptor) pair\n",
    "    groups = np.array([f\"{b}_{p}\" for b, p in zip(groups_receptors, groups_phage)])\n",
    "else:\n",
    "    raise ValueError(\"Invalid grouping strategy selected.\")\n",
    "\n",
    "print(f\"Number of unique groups: {len(set(groups))}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d6399087",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/5 [00:00<?, ?it/s]/home/snt/.pyenv/versions/yakuza_venv/lib/python3.10/site-packages/xgboost/training.py:183: UserWarning: [20:14:31] WARNING: /workspace/src/learner.cc:738: \n",
      "Parameters: { \"use_label_encoder\" } are not used.\n",
      "\n",
      "  bst.update(dtrain, iteration=i, fobj=obj)\n",
      "100%|██████████| 5/5 [00:00<00:00, 14.63it/s]\n"
     ]
    }
   ],
   "source": [
    "# Convert group names to numeric codes (required for splitting)\n",
    "unique_groups = list(set(groups))\n",
    "group_to_id = {g: i for i, g in enumerate(unique_groups)}\n",
    "group_ids = np.array([group_to_id[g] for g in groups])\n",
    "\n",
    "k = 5\n",
    "gkf = GroupKFold(n_splits=k)\n",
    "results = []\n",
    "true_labels = []\n",
    "\n",
    "for train_idx, test_idx in tqdm(gkf.split(X_full, y_full, groups=group_ids), total=k):\n",
    "    X_train, X_test = X_full[train_idx], X_full[test_idx]\n",
    "    y_train, y_test = y_full[train_idx], y_full[test_idx]\n",
    "\n",
    "    imbalance = np.sum(y_train == 1) / np.sum(y_train == 0)\n",
    "    best_config[\"scale_pos_weight\"] = 1 / imbalance\n",
    "    \n",
    "    xgb = XGBClassifier(\n",
    "        **best_config,\n",
    "        n_jobs=8,\n",
    "        eval_metric='logloss',\n",
    "        use_label_encoder=False\n",
    "    )\n",
    "    xgb.fit(X_train, y_train)\n",
    "    y_score = xgb.predict_proba(X_test)[:, 1]\n",
    "\n",
    "    results.append(y_score)\n",
    "    true_labels.append(y_test)\n",
    "\n",
    "# Save results\n",
    "with open(f\"{results_path}/serr_receptor_level_kfold_results.pickle\", \"wb\") as f:\n",
    "    pickle.dump({'scores': results, 'labels': true_labels}, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "68d60c57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([array([0.63587373, 0.36634678, 0.63587373, 0.36634678], dtype=float32),\n",
       "  array([0.50473005, 0.39629772, 0.6089619 , 0.49906892], dtype=float32),\n",
       "  array([0.50473005, 0.39629772, 0.6089619 , 0.49906892], dtype=float32),\n",
       "  array([0.50263625, 0.39459795, 0.60965633, 0.5001147 ], dtype=float32),\n",
       "  array([0.63587373, 0.36634678, 0.63587373, 0.36634678], dtype=float32)],\n",
       " [array([0, 0, 1, 1]),\n",
       "  array([1, 0, 1, 0]),\n",
       "  array([1, 0, 1, 0]),\n",
       "  array([1, 0, 1, 0]),\n",
       "  array([0, 0, 1, 1])])"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results, true_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5edf47ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "yakuza_venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
